{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# default_exp sequenceplot"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Sequence visualization"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#hide\n",
    "import sys\n",
    "sys.path.append('../')\n",
    "from alphamap.uniprot_integration import uniprot_feature_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#export\n",
    "import os\n",
    "import tempfile"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Format uniprot annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#export\n",
    "import pandas as pd\n",
    "\n",
    "def format_uniprot_annotation(uniprot_ann: pd.DataFrame, uniprot_feature_dict: dict):\n",
    "    \"\"\"\n",
    "    Function to format uniprot annotation for plotting.\n",
    "\n",
    "    Args:\n",
    "        uniprot_ann (pd.DataFrame): Formatted uniprot annotations from alphamap.\n",
    "        uniprot_feature_dict (dict): Uniprot feature dictionary defined by alphamap.\n",
    "    Returns:\n",
    "        pd.DataFrame: Uniprot annotation with a combined structure entry for helix, strand and turn.\n",
    "\n",
    "    \"\"\"\n",
    "    uniprot = uniprot_ann.copy(deep=True)\n",
    "    uniprot.loc[uniprot.feature == \"HELIX\", \"note\"] = \"Helix\"\n",
    "    uniprot.loc[uniprot.feature == \"STRAND\", \"note\"] = \"Beta strand\"\n",
    "    uniprot.loc[uniprot.feature == \"TURN\", \"note\"] = \"Turn\"\n",
    "    uniprot.loc[uniprot.feature.isin([\"HELIX\",\"STRAND\",\"TURN\"]), \"feature\"] = \"STRUCTURE\"\n",
    "\n",
    "    uniprot_feature_dict_rev = {v: k for k, v in uniprot_feature_dict.items()}\n",
    "\n",
    "    uniprot['annotation'] = uniprot['note']\n",
    "    uniprot.loc[uniprot['annotation'].isnull(), 'annotation'] = uniprot['feature']\n",
    "    uniprot = uniprot.replace({\"annotation\": uniprot_feature_dict_rev})\n",
    "    return uniprot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#hide\n",
    "import numpy as np\n",
    "\n",
    "def test_format_uniprot_annotation():\n",
    "    in_df = pd.read_csv('../alphamap/data/preprocessed_uniprot_human.csv')\n",
    "    id_df_structure = in_df[in_df.protein_id  == \"P43166\"]\n",
    "    id_df_structure_form = format_uniprot_annotation(id_df_structure, uniprot_feature_dict)\n",
    "    \n",
    "    np.testing.assert_equal(np.unique(id_df_structure_form[id_df_structure_form.feature==\"STRUCTURE\"].annotation), \n",
    "                            ['Beta strand', 'Helix', 'Turn'])\n",
    "    \n",
    "    id_df_nan = in_df[in_df.protein_id  == \"Q8IYX3\"]\n",
    "    id_df_nan_form = format_uniprot_annotation(id_df_nan, uniprot_feature_dict)\n",
    "    np.testing.assert_equal(id_df_nan_form[id_df_nan_form.feature==\"COILED\"].annotation.values,\n",
    "                            'Coiled coil')\n",
    "    \n",
    "test_format_uniprot_annotation()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#hide\n",
    "\n",
    "###### Input data for testing ###### \n",
    "\n",
    "import pandas as pd\n",
    "from pyteomics import fasta\n",
    "\n",
    "test_formatted_proteome_data = pd.DataFrame(data={'unique_protein_id': [\"A0A024R161\", \"A0A087WT10\", \"A0A087WTH1\", \n",
    "                                                                             \"A0A024R161\", \"A0A087WT10\", \n",
    "                                                                             \"A0A087WTH5\"], \n",
    "                                                       'modified_sequence': [\"PEPT[Phospho (STY)]IDER\", \"PEPT[Phospho (STY)]IDER\", \"PEPT[Phospho (STY)]IDER\",\n",
    "                                                                             \"SEQ[GlyGly (K)]UENCE[GlyGly (K)]R\", \"SEQ[GlyGly (K)]UENCE[GlyGly (K)]R\", \n",
    "                                                                             \"VIEWER\"],\n",
    "                                                       'naked_sequence': [\"PEPTIDER\", \"PEPTIDER\", \"PEPTIDER\", \n",
    "                                                                          \"SEQUENCER\", \"SEQUENCER\", \n",
    "                                                                          \"VIEWER\"],\n",
    "                                                  'all_protein_ids': [\"A0A024R161;A0A087WT10;A0A087WTH1\", \"A0A024R161;A0A087WT10;A0A087WTH1\", \"A0A024R161;A0A087WT10;A0A087WTH1\", \n",
    "                                                                             \"A0A024R161;A0A087WT10\", \"A0A024R161;A0A087WT10\", \n",
    "                                                                             \"A0A087WTH5\"],\n",
    "                                                       'start':[3,28,107,95,150,1],\n",
    "                                                       'end':[10,35,114,103,158,6], \n",
    "                                           'PTMsites':[[3],[3],[3],[2,7],[2,7],[]],\n",
    "                                           'PTMtypes':[[\"[Phospho (STY)]\"],[\"[Phospho (STY)]\"],[\"[Phospho (STY)]\"],[\"[GlyGly (K)]\",\"[GlyGly (K)]\"],[\"[GlyGly (K)]\",\"[GlyGly (K)]\"],[]]})\n",
    "\n",
    "test_fasta = fasta.IndexedUniProt(\"../testdata/test.fasta\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# export \n",
    "\n",
    "# The PTM shape dictionary defines the shapes for each modification, so they are consistent across datasets and analyses.\n",
    "\n",
    "ptm_shape_dict = {\n",
    "    '[Acetyl (K)]': 5,\n",
    "    '[Acetyl (Protein N-term)]': 5,\n",
    "    '[Carbamidomethyl (C)]': 3,\n",
    "    '[Oxidation (M)]': 4,\n",
    "    '[Phospho (STY)]': 0,\n",
    "    '[GlyGly (K)]': 2,\n",
    "    '[Methyl (KR)]': 6,\n",
    "    '[Dimethyl (KR)]': 6,\n",
    "    '[Trimethyl (K)]': 6,\n",
    "    '[Pro5]': 9,\n",
    "    '[Pro6]': 10,\n",
    "    '[Glu->pyro-Glu]': 11,\n",
    "    '[Gln->pyro-Glu]': 12,\n",
    "    '[QQTGG (K)]': 13,\n",
    "    '[Deamidation (N)]': 14,\n",
    "    '[Deamidation 18O (N)]': 14,\n",
    "    '[Deamidation (NQ)]': 14,\n",
    "    '[Hydroxyproline]': 15,\n",
    "    '[Carbamyl (N-term)]': 16,\n",
    "    '[Delta:H(2)C(2) (N-term)]': 19,\n",
    "    '[Dioxidation (MW)]': 4,\n",
    "    '[Trioxidation (C)]': 4,\n",
    "    '[Dethiomethyl (M)]': 20,\n",
    "    '[Cation:Na (DE)]': 21,\n",
    "    '[Methyl (E)]': 6,\n",
    "    '[Dehydrated (ST)]': 23,\n",
    "    '[Oxidation (P)]': 4,\n",
    "    '[Dimethyl (K)]': 6,\n",
    "    '[Amidated (Protein C-term)]': 22,\n",
    "    '[Sulfo (STY)]': 24,\n",
    "    '[Acetyl (N-term)]': 5,\n",
    "    '[Amidated (C-term)]': 22,\n",
    "    '[Sulfation (Y)]': 25,\n",
    "    '[Phospho (ST)]': 0,\n",
    "    '[Cys-Cys]': 26,\n",
    "    '[Cysteinyl]': 27,\n",
    "    '[Cysteinyl - carbamidomethyl]': 28,\n",
    "    '[Oxidation (MP)]': 4\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Format data for plotting"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#export \n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from pyteomics import fasta\n",
    "\n",
    "def get_plot_data(protein,df,fasta):\n",
    "    \"\"\"\n",
    "    Function to format experimental data for plotting.\n",
    "\n",
    "    Args:\n",
    "        protein (str): Uniprot protein accession.\n",
    "        df (pd.DataFrame): Experimental data imported and formatted according to alphamap standards.\n",
    "        fasta (fasta): Fasta file imported by pyteomics 'fasta.IndexedUniProt'.\n",
    "    Returns:\n",
    "        pd.DataFrame: Formatted dataframe for plotting.\n",
    "\n",
    "    \"\"\"\n",
    "    protein_sequence = fasta[protein].sequence\n",
    "    df_prot = df[df.unique_protein_id==protein]\n",
    "\n",
    "    if df_prot.shape[0] == 0:\n",
    "        df_plot = None\n",
    "    else:\n",
    "        df_peps = [np.arange(row['start'], row['end']+1) for _, row in df_prot.iterrows()]\n",
    "        df_peps  = pd.DataFrame.from_records(data=df_peps)\n",
    "        df_peps['modified_sequence'] = df_prot['modified_sequence'].values\n",
    "        df_peps['all_protein_ids'] = df_prot['all_protein_ids'].values\n",
    "        df_peps = df_peps.melt(id_vars=['modified_sequence','all_protein_ids'])\n",
    "        df_peps = df_peps[['modified_sequence','all_protein_ids','value']].dropna()\n",
    "        df_peps = df_peps.rename(columns={\"value\": \"seq_position\"})\n",
    "        df_peps['marker_symbol'] = 1\n",
    "        df_peps['marker_size'] = 8\n",
    "        df_peps['PTM'] = np.nan\n",
    "        df_peps['PTMtype'] = np.nan\n",
    "        df_peps['PTMshape'] = np.nan\n",
    "        unique_pep = df_peps.modified_sequence.unique()\n",
    "        for uid in unique_pep:\n",
    "            df_peps_uid = df_peps[df_peps.modified_sequence==uid]\n",
    "            start_uid = np.min(df_peps_uid.seq_position)\n",
    "            end_uid = np.max(df_peps_uid.seq_position)\n",
    "            df_peps['marker_symbol'] = np.where(df_peps.seq_position == start_uid, 7, df_peps.marker_symbol)\n",
    "            df_peps['marker_symbol'] = np.where(df_peps.seq_position == end_uid, 8, df_peps.marker_symbol)\n",
    "            df_peps['marker_size'] = np.where(df_peps.seq_position == start_uid, 6, df_peps.marker_size)\n",
    "            df_peps['marker_size'] = np.where(df_peps.seq_position == end_uid, 6, df_peps.marker_size)\n",
    "\n",
    "            df_PTMs_uid = df_prot[df_prot.modified_sequence==uid]\n",
    "            #PTMsites = df_PTMs_uid.PTMsites.tolist()[0] + start_uid\n",
    "            PTMsites = [x+start_uid for x in df_PTMs_uid.PTMsites.tolist()[0]]\n",
    "            PTMtypes = df_PTMs_uid.PTMtypes.tolist()[0]\n",
    "            \n",
    "            #print(PTMsites)\n",
    "            #print(PTMtypes)\n",
    "\n",
    "            for i in range(0,len(PTMsites)):\n",
    "                df_peps['PTM'] = np.where(df_peps[\"seq_position\"]==PTMsites[i], 1, df_peps.PTM)\n",
    "                df_peps['PTMtype'] = np.where(df_peps[\"seq_position\"]==PTMsites[i], PTMtypes[i], df_peps.PTMtype)\n",
    "\n",
    "            #print(df_peps)\n",
    "            \n",
    "            df_seq = pd.DataFrame({'seq_position':np.arange(0,len(protein_sequence))})\n",
    "\n",
    "            df_plot = pd.merge(df_seq, df_peps, how='left', on='seq_position')\n",
    "            df_plot['height']=0\n",
    "            df_plot['color']=\"grey\"\n",
    "\n",
    "            unique_mods = df_plot['PTMtype'].dropna().unique()\n",
    "            if len(unique_mods) > 0:\n",
    "                for mod in df_plot['PTMtype'].dropna().unique():\n",
    "                    if mod != 'nan':\n",
    "                        #print(mod)\n",
    "                        if mod not in ptm_shape_dict.keys():\n",
    "                            ptm_shape_dict.update({mod : 17})\n",
    "\n",
    "                        df_plot.loc[df_plot.PTMtype == mod, 'PTMshape'] = ptm_shape_dict[mod]\n",
    "                        #df_plot.loc[df_plot.PTMtype == mod, 'PTMshape'] = 17\n",
    "            \n",
    "        #print(df_plot)\n",
    "\n",
    "        df_plot['modified_sequence']= df_plot['modified_sequence'].astype(str)\n",
    "        df_plot = df_plot.groupby('seq_position').agg({'modified_sequence':'; '.join, \n",
    "                                                       'all_protein_ids': 'first', \n",
    "                                                       'marker_symbol': 'first',\n",
    "                                                       'marker_size': 'first',\n",
    "                                                       'PTM': 'first',\n",
    "                                                       'PTMtype': 'first',\n",
    "                                                       'PTMshape': 'first',\n",
    "                                                       'height': 'first',\n",
    "                                                       'color': 'first'\n",
    "                                                      }).reset_index()\n",
    "        \n",
    "        #print(df_plot)\n",
    "\n",
    "    return(df_plot)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#hide\n",
    "#@ToDo: write tests for get_plot_data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Function to plot a single dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#export \n",
    "import plotly.graph_objects as go\n",
    "\n",
    "def plot_single_peptide_traces(df_plot,protein,fasta):\n",
    "    \"\"\"\n",
    "    Function to plot single peptide trace.\n",
    "\n",
    "    Args:\n",
    "        df_plot (pd.DataFrame): Formatted dataframe for plotting, generated by get_plot_data.\n",
    "        protein (str): Uniprot protein accession.\n",
    "        fasta (fasta): Fasta file imported by pyteomics 'fasta.IndexedUniProt'.\n",
    "    Returns:\n",
    "        go.Figure: Figure data for a single dataset.\n",
    "\n",
    "    \"\"\"\n",
    "    protein_sequence = fasta[protein].sequence\n",
    "    entry_name = fasta[protein].description['GN']\n",
    "    protein_name = fasta[protein].description['name']\n",
    "\n",
    "    plot0 = go.Scatter(y=[None],\n",
    "                       name='',\n",
    "                       xaxis='x1',\n",
    "                       showlegend=False)\n",
    "\n",
    "    ## Peptide backbone\n",
    "    df_plot_pep = df_plot.dropna(subset=['modified_sequence'])\n",
    "    df_plot_pep = df_plot_pep[~df_plot_pep.modified_sequence.str.contains('nan')]\n",
    "    plot1 = go.Scatter(x=df_plot_pep.seq_position+1,\n",
    "                       y=df_plot.height,\n",
    "                       xaxis='x2',\n",
    "                       mode='markers',\n",
    "                       marker_size=df_plot_pep.marker_size,\n",
    "                       marker_symbol=df_plot_pep.marker_symbol,\n",
    "                       marker_line_color=df_plot_pep.color,\n",
    "                       marker_color=df_plot_pep.color,\n",
    "                       marker_opacity=1,\n",
    "                       meta=df_plot_pep.modified_sequence,\n",
    "                       text=df_plot_pep.all_protein_ids,\n",
    "                       hovertemplate ='Peptide: %{meta}<br>' +\n",
    "                       'Protein IDs: %{text}',\n",
    "                       name='',\n",
    "                       showlegend=False)\n",
    "\n",
    "    covered_AA = len(df_plot_pep.seq_position.unique())\n",
    "    percent_AA_coverage = int(np.round(100/len(protein_sequence)*covered_AA))\n",
    "    #print(percent_AA_coverage)\n",
    "\n",
    "    ## PTM dots\n",
    "    df_plot_ptm = df_plot.dropna(subset=['PTM'])\n",
    "    #print(df_plot_ptm)\n",
    "    plot2 = go.Scatter(x=df_plot_ptm.seq_position+1,\n",
    "                       y=df_plot_ptm.height+0.3,\n",
    "                       xaxis='x2',\n",
    "                       mode='markers',\n",
    "                       marker_size=8,\n",
    "                       marker_symbol=df_plot_ptm.PTMshape,\n",
    "                       marker_line_color=df_plot_ptm.color,\n",
    "                       marker_color=df_plot_ptm.color,\n",
    "                       marker_opacity=1,\n",
    "                       text=df_plot_ptm.PTMtype,\n",
    "                       hovertemplate = 'PTM: %{text}',\n",
    "                       #hoverinfo='text',\n",
    "                       name='',\n",
    "                       showlegend=False)\n",
    "\n",
    "    layout = go.Layout(\n",
    "            yaxis=dict(\n",
    "                title = \"\",\n",
    "                ticks = None,\n",
    "                showticklabels=False,\n",
    "                range=[-1, 2],\n",
    "                showgrid=False,\n",
    "                zeroline=False\n",
    "                ),\n",
    "            xaxis1=dict(\n",
    "                title= 'protein sequence',\n",
    "                tickmode = 'array',\n",
    "                range=[-10, len(protein_sequence)+10],\n",
    "                tickvals = np.arange(1,len(protein_sequence)+1),\n",
    "                ticktext = list(protein_sequence),\n",
    "                tickangle=0,\n",
    "                matches=\"x2\",\n",
    "                type=\"linear\",\n",
    "                anchor=\"y\",\n",
    "                showgrid=False,\n",
    "                zeroline=False\n",
    "                ),\n",
    "            xaxis2=dict(\n",
    "                title= 'AA position',\n",
    "                tickmode = 'auto',\n",
    "                range=[-10, len(protein_sequence)+10],\n",
    "                tickangle=0,\n",
    "                matches=\"x1\",\n",
    "                side=\"top\",\n",
    "                type=\"linear\",\n",
    "                anchor=\"y\",\n",
    "                showgrid=False,\n",
    "                zeroline=False,\n",
    "                tickformat = '.d'\n",
    "                ),\n",
    "        #showlegend=False,\n",
    "        #height=400,\n",
    "        #width=1000,\n",
    "        plot_bgcolor='rgba(0,0,0,0)',\n",
    "        title=f\"Sequence plot for: {protein_name}<br>{entry_name} - {protein}\",\n",
    "        meta=percent_AA_coverage,\n",
    "        margin = dict(l=20, r=20, t=150, b=20)\n",
    "        )\n",
    "\n",
    "    fig = go.Figure(data=[plot1,plot2,plot0], layout=layout)\n",
    "\n",
    "    #print(fig.layout.meta)\n",
    "\n",
    "    for i in range(0, df_plot_ptm.shape[0]):\n",
    "            fig.add_shape(\n",
    "                    dict(\n",
    "                        type=\"line\",\n",
    "                        xref=\"x2\",\n",
    "                        x0=df_plot_ptm.seq_position.values[i] +1,\n",
    "                        y0=df_plot_ptm.height.values[i],\n",
    "                        x1=df_plot_ptm.seq_position.values[i] +1,\n",
    "                        y1=df_plot_ptm.height.values[i]+0.3,\n",
    "                        line=dict(\n",
    "                            color=df_plot_ptm.color.values[i],\n",
    "                            width=1\n",
    "                        )\n",
    "                    )\n",
    "            )\n",
    "\n",
    "    return fig"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#export\n",
    "\n",
    "# Color palettes used for the different uniprot annotation features\n",
    "custom_color_palettes = {\n",
    "    'col_greens':[\"#5C965D\",\"#6AA16B\",\"#77AC78\",\"#84B786\",\"#91C193\",\"#9FCCA1\",\"#B3DCB5\",\"#C6EBC9\"],\n",
    "    'col_ornages':[\"#ff4800\",\"#ff5400\",\"#ff6000\",\"#ff6d00\",\"#ff7900\",\"#ff8500\",\"#ff9100\",\"#ff9e00\",\"#ffaa00\",\"#ffb600\"],\n",
    "    'col_purples':[\"#ffa69e\",\"#febaae\",\"#fcb088\",\"#d9f3e2\",\"#b8f2e6\",\"#aed9e0\",\"#9baed9\",\"#9199d5\",\"#8783d1\"],\n",
    "    'col_turquises':[\"#00a9a5\",\"#4e8098\",\"#90c2e7\"],\n",
    "    'col_darkpinks':[\"#42033d\",\"#6f0c59\",\"#901468\",\"#7c238c\",\"#924ea6\",\"#9c5eae\"],\n",
    "    'col_browns':[\"#5a2a27\",\"#5c4742\",\"#8d5b4c\",\"#a5978b\",\"#c4bbaf\"]\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#export\n",
    "\n",
    "# The uniprot_color_dict maps each uniprot annotation feature to a color in the custom_color_palettes.\n",
    "uniprot_color_dict = {'CHAIN': custom_color_palettes['col_greens'][0],\n",
    "                      'INIT_MET': custom_color_palettes['col_greens'][1],\n",
    "                      'PEPTIDE': custom_color_palettes['col_greens'][2],\n",
    "                      'PROPEP': custom_color_palettes['col_greens'][3],\n",
    "                      'SIGNAL': custom_color_palettes['col_greens'][4],\n",
    "                      'TRANSIT': custom_color_palettes['col_greens'][5],\n",
    "\n",
    "                      'COILED': custom_color_palettes['col_purples'][0],\n",
    "                      'COMPBIAS': custom_color_palettes['col_purples'][1],\n",
    "                      'DOMAIN': custom_color_palettes['col_purples'][2],\n",
    "                      'MOTIF': custom_color_palettes['col_purples'][3],\n",
    "                      'REGION': custom_color_palettes['col_purples'][4],\n",
    "                      'REPEAT': custom_color_palettes['col_purples'][5],\n",
    "                      'ZN_FING': custom_color_palettes['col_purples'][6],\n",
    "\n",
    "                      'INTRAMEM': custom_color_palettes['col_turquises'][0],\n",
    "                      'TOPO_DOM': custom_color_palettes['col_turquises'][1],\n",
    "                      'TRANSMEM': custom_color_palettes['col_turquises'][2],\n",
    "\n",
    "                      'STRUCTURE': 'black',\n",
    "                      # extra structures\n",
    "                      'Helix': '#5dabe8',\n",
    "                      'Turn': '#e094bc',\n",
    "                      'Beta strand': '#8cdbad',\n",
    "\n",
    "                      'CROSSLNK': custom_color_palettes['col_ornages'][2],\n",
    "                      'DISULFID': custom_color_palettes['col_ornages'][3],\n",
    "                      'CARBOHYD': custom_color_palettes['col_ornages'][4],\n",
    "                      'LIPID': custom_color_palettes['col_ornages'][5],\n",
    "                      'MOD_RES': custom_color_palettes['col_ornages'][6],\n",
    "\n",
    "                      'BINDING': custom_color_palettes['col_darkpinks'][0],\n",
    "                      'CA_BIND': custom_color_palettes['col_darkpinks'][1],\n",
    "                      'DNA_BIND': custom_color_palettes['col_darkpinks'][2],\n",
    "                      'METAL': custom_color_palettes['col_darkpinks'][3],\n",
    "                      'NP_BIND': custom_color_palettes['col_darkpinks'][4],\n",
    "                      'SITE': custom_color_palettes['col_darkpinks'][5],\n",
    "\n",
    "                      'NON_STD': custom_color_palettes['col_browns'][0],\n",
    "                      'NON_CONS': custom_color_palettes['col_browns'][1],\n",
    "                      'NON_TER': custom_color_palettes['col_browns'][2],\n",
    "                      'VARIANT': custom_color_palettes['col_browns'][3],\n",
    "                      'CONFLICT': custom_color_palettes['col_browns'][4],\n",
    "\n",
    "                      'VAR_SEQ': '#fae7b1',\n",
    "                      'UNSURE': 'grey',\n",
    "                      'MUTAGEN': 'darkgrey',\n",
    "                     }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#export\n",
    "\n",
    "# Dictionary that maps one-letter amino acid abbreviations to their full name.\n",
    "aa_color_dict = {'A':'Alanine',\n",
    "                 'R':'Arginine',\n",
    "                 'N':'Asparagine',\n",
    "                 'D':'Aspartic acid',\n",
    "                 'C':'Cysteine',\n",
    "                 'E':'Glutamic acid',\n",
    "                 'Q':'Glutamine',\n",
    "                 'G':'Glycine',\n",
    "                 'H':'Histidine',\n",
    "                 'I':'Isoleucine',\n",
    "                 'L':'Leucine',\n",
    "                 'K':'Lysine',\n",
    "                 'M':'Methionine',\n",
    "                 'F':'Phenylalanine',\n",
    "                 'P':'Proline',\n",
    "                 'S':'Serine',\n",
    "                 'T':'Threonine',\n",
    "                 'W':'Tryptophan',\n",
    "                 'Y':'Tyrosine',\n",
    "                 'V':'Valine',\n",
    "                 'X':'nan',\n",
    "                 'U':'nan'}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Plotting function for the full sequence plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#export\n",
    "from structuremap.processing import download_alphafold_cif, download_alphafold_pae, format_alphafold_data, annotate_accessibility, get_smooth_score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#export\n",
    "def get_quality_category(s):\n",
    "    if s >= 90:\n",
    "        n='Very high'\n",
    "    elif s >= 70:\n",
    "        n='Confident'\n",
    "    elif s >= 50:\n",
    "        n='Low'\n",
    "    else:\n",
    "        n='Very low'\n",
    "    return(n)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#export\n",
    "def get_exposure_category(s):\n",
    "    if s > 10:\n",
    "        n='Surrounded'\n",
    "    elif s > 5:\n",
    "        n='Many neighbors'\n",
    "    elif s > 0:\n",
    "        n='Few neighbors'\n",
    "    else:\n",
    "        n='No neighbors'\n",
    "    return(n)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#export\n",
    "def get_alphafold_annotation(protein: str, \n",
    "                             selected_features: list, \n",
    "                             download_folder: str = tempfile.gettempdir()) -> pd.DataFrame:\n",
    "    \n",
    "    alphafold_feature_dict = dict({'AlphaFold confidence':'quality', \n",
    "        'AlphaFold exposure':'nAA_12_70_pae', \n",
    "        'AlphaFold IDR':'AlphaFold IDR',\n",
    "        'AlphaFold secondary structures':'structure_group'})\n",
    "    \n",
    "    alphafold_features = [i for i in [\n",
    "        'AlphaFold confidence', \n",
    "        'AlphaFold exposure', \n",
    "        'AlphaFold IDR',\n",
    "        'AlphaFold secondary structures'\n",
    "    ] if i in selected_features]\n",
    "\n",
    "    if len(alphafold_features) > 0:\n",
    "\n",
    "        # download cif file\n",
    "        valid_proteins, invalid_proteins, existing_proteins = download_alphafold_cif(\n",
    "            proteins=[protein],\n",
    "            out_folder=download_folder)\n",
    "        is_available = np.sum([len(valid_proteins),len(existing_proteins)])\n",
    "        \n",
    "        if is_available > 0:\n",
    "\n",
    "            af_data = format_alphafold_data(directory = download_folder, protein_ids = [protein])\n",
    "\n",
    "            if len([i for i in ['AlphaFold exposure',\n",
    "                                'AlphaFold IDR'] if i in selected_features]) > 0:\n",
    "                # download pae file\n",
    "                valid_pae, invalid_pae, existing_pae = download_alphafold_pae(\n",
    "                    proteins=[protein],\n",
    "                    out_folder=download_folder)\n",
    "                is_available_pae = np.sum([len(valid_pae),len(existing_pae)])\n",
    "                #@ToDo\n",
    "                #if is_available == 0:\n",
    "                #    print('error')\n",
    "\n",
    "                # annotate exposure\n",
    "                exposure_df = annotate_accessibility(df = af_data,\n",
    "                                                     max_dist = 12,\n",
    "                                                     max_angle = 70,\n",
    "                                                     error_dir = download_folder)\n",
    "\n",
    "                af_data = af_data.merge(exposure_df, how='left', on=['protein_id','AA','position'])\n",
    "                \n",
    "                # annotate exposure\n",
    "                full_exposure_df = annotate_accessibility(df = af_data,\n",
    "                                                          max_dist = 24,\n",
    "                                                          max_angle = 180,\n",
    "                                                          error_dir = download_folder)\n",
    "\n",
    "                af_data = af_data.merge(full_exposure_df, how='left', on=['protein_id','AA','position'])\n",
    "\n",
    "            # smooth scores\n",
    "            smoothing_scores = []\n",
    "            if len([i for i in ['AlphaFold IDR'] if i in alphafold_features]) > 0:\n",
    "                smoothing_scores.append('nAA_24_180_pae')\n",
    "            if len(smoothing_scores) > 0:     \n",
    "                af_data = get_smooth_score(af_data, np.array(smoothing_scores),[10])\n",
    "\n",
    "            # annotate predicted IDRs\n",
    "            if 'AlphaFold IDR' in alphafold_features:\n",
    "                af_data['AlphaFold IDR'] = np.where(af_data['nAA_24_180_pae_smooth10']<=34.27, \"IDR\", \"folded\")\n",
    "\n",
    "            af_columns = ['position','AA']\n",
    "            af_columns = af_columns + [alphafold_feature_dict[i] for i in alphafold_features]\n",
    "            af_data = af_data[af_columns]\n",
    "\n",
    "            if 'quality' in af_data.columns:\n",
    "                af_data[\"AlphaFold confidence\"] = [get_quality_category(x) for x in af_data.quality] \n",
    "\n",
    "            if 'nAA_12_70_pae' in af_data.columns:\n",
    "                af_data[\"AlphaFold exposure\"] = [get_exposure_category(x) for x in af_data.nAA_12_70_pae] \n",
    "                \n",
    "            if 'structure_group' in af_data.columns:\n",
    "                af_data = af_data.rename(columns={\"structure_group\": \"AlphaFold secondary structures\"})\n",
    "\n",
    "        else:\n",
    "            af_data = pd.DataFrame({'AA' : []})\n",
    "    else:\n",
    "        af_data = pd.DataFrame({'AA' : []})\n",
    "        \n",
    "    return(af_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# #hide\n",
    "# af_annotation = get_alphafold_annotation(protein= \"Q00266\", \n",
    "#                              selected_features= [\"AlphaFold confidence\",  \n",
    "#                                                    \"AlphaFold exposure\", \n",
    "#                                                    \"AlphaFold IDR\", \n",
    "#                                                    \"AlphaFold secondary structures\"], \n",
    "#                              download_folder= '/Users/isabell/Downloads/') "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#export \n",
    "\n",
    "import plotly.graph_objects as go\n",
    "from alphamap.proteolytic_cleavage import get_cleavage_sites\n",
    "\n",
    "def plot_peptide_traces(df: pd.DataFrame or list,\n",
    "                        name: str or list,\n",
    "                        protein: str,\n",
    "                        fasta: fasta,\n",
    "                        uniprot: pd.DataFrame,\n",
    "                        selected_features: list,\n",
    "                        uniprot_feature_dict: dict,\n",
    "                        uniprot_color_dict: dict,\n",
    "                        selected_proteases: list = [],\n",
    "                        selected_alphafold_features: list = [],\n",
    "                        dashboard: bool = False,\n",
    "                        trace_colors: list = [],\n",
    "                        download_folder: str = tempfile.gettempdir()):\n",
    "\n",
    "    \"\"\"\n",
    "    Function to generate the sequence plot.\n",
    "\n",
    "    Args:\n",
    "        df (pd.DataFrame/list): Single dataframe or list of dataframes containing the datasets to plot.\n",
    "        name (str/list): Single string or list of strings containing the names for each dataset in df.\n",
    "        protein (str): Uniprot protein accession.\n",
    "        fasta (fasta): Fasta file imported by pyteomics 'fasta.IndexedUniProt'.\n",
    "        uniprot (pd.DataFrame): Uniprot annotations formatted by alphamap.\n",
    "        selected_features (list): List of uniprot features to plot.\n",
    "        uniprot_feature_dict (dict): Uniprot feature dictionary.\n",
    "        uniprot_color_dict (dict): Uniprot color dictionary.\n",
    "        selected_proteases (list, optional): List of proteases to plot. Default is an empty list.\n",
    "        dashboard (bool, optional): Flag if the function is called from the dashboard. Default is 'False'.\n",
    "        trace_colors (list, optional): List of manualy selected colors for each dataset in df. Default is an empty list.\n",
    "\n",
    "    Returns:\n",
    "        go.Figure: Sequence plot.\n",
    "\n",
    "    \"\"\"\n",
    "\n",
    "    figure_height = 200\n",
    "\n",
    "    protein_sequence = fasta[protein].sequence\n",
    "\n",
    "    # colors for experimental data traces\n",
    "    if len(trace_colors) == 0:\n",
    "        colors = [\"#023e8a\",\"#0096c7\",\"#90e0ef\",\"#7fd14d\",\"#26a96c\",\n",
    "                  \"#0D7F8B\", \"#24B4AB\", \"#9DE7BE\", \"#1B9CC6\", \"#016699\"]\n",
    "    else:\n",
    "        colors = trace_colors\n",
    "\n",
    "    # generation of a reverse uniprot_feature_dict\n",
    "    uniprot_feature_dict_rev = {v: k for k, v in uniprot_feature_dict.items()}\n",
    "    #uniprot_feature_dict_rev[\"STRUCTURE\"] = \"Secondary structure\"\n",
    "\n",
    "    # subsetting of the uniprot annotation to the selected features\n",
    "    uniprot_annotation_p = uniprot[uniprot.protein_id==protein]\n",
    "    # formatting of uniprot annotations\n",
    "    uniprot_annotation_p_f = format_uniprot_annotation(uniprot_annotation_p, uniprot_feature_dict)\n",
    "    # subset for selected features\n",
    "    uniprot_annotation_p_f_f = uniprot_annotation_p_f[uniprot_annotation_p_f.feature.isin(selected_features)]\n",
    "\n",
    "    #@ToDo if \n",
    "    \n",
    "    if isinstance(df, pd.DataFrame):\n",
    "        df_plot = get_plot_data(protein=protein,\n",
    "                              df = df,\n",
    "                              fasta = fasta)\n",
    "\n",
    "        df_plot.color = colors[0]\n",
    "\n",
    "        observed_mods = list(set(df_plot.PTMtype))\n",
    "        ptm_shape_dict_sub = {key: ptm_shape_dict[key] for key in observed_mods if key in ptm_shape_dict}\n",
    "\n",
    "        fig = plot_single_peptide_traces(df_plot,protein=protein,fasta = fasta)\n",
    "\n",
    "        AA_coverage = fig.layout.meta\n",
    "        trace_name = [name + \"<br> (\" + str(AA_coverage) + \"% coverage)\"]\n",
    "\n",
    "        fig.update_layout(yaxis=dict(showticklabels=True,\n",
    "                                     tickmode = 'array',\n",
    "                                     tickvals = [0],\n",
    "                                     ticktext = [name + \"(\" + str(AA_coverage) + \"%)\"],\n",
    "                                     showgrid=False))\n",
    "\n",
    "        y_max = 1\n",
    "\n",
    "    elif isinstance(df, list):\n",
    "\n",
    "        df_plot = [get_plot_data(protein=protein,\n",
    "                               df = d,\n",
    "                               fasta = fasta) for d in df]\n",
    "\n",
    "        # Subset data and annotations for the samples where the selected protein was detected\n",
    "        valid_idx = []\n",
    "        for i in range(len(df_plot)):\n",
    "            if df_plot[i] is not None:\n",
    "                valid_idx.append(i)\n",
    "        df_plot = [df_plot[i] for i in valid_idx]\n",
    "        name = [name[i] for i in valid_idx]\n",
    "        colors = [colors[i] for i in valid_idx]\n",
    "        #observed_mods = set([df_plot[i].PTMtype for i in valid_idx])\n",
    "        observed_mods = []\n",
    "        for i in range(len(df_plot)):\n",
    "            observed_mods.extend(list(set(df_plot[i].PTMtype)))\n",
    "        observed_mods = list(set(observed_mods))\n",
    "        ptm_shape_dict_sub = {key: ptm_shape_dict[key] for key in observed_mods if key in ptm_shape_dict}\n",
    "\n",
    "        for i in range(len(df_plot)):\n",
    "            df_plot[i].color = colors[i]\n",
    "            df_plot[i].height = 1+i\n",
    "\n",
    "        plot_list = [plot_single_peptide_traces(df,protein=protein,fasta = fasta) for df in df_plot]\n",
    "        new_data = [p.data for p in plot_list]\n",
    "        new_data = sum(new_data, ())\n",
    "        new_layout = plot_list[0].layout\n",
    "        shapes = [p.layout.shapes for p in plot_list]\n",
    "        shapes = sum(shapes, ())\n",
    "        new_layout.shapes = new_layout.shapes + tuple(shapes)\n",
    "        AA_coverage = [p.layout.meta for p in plot_list]\n",
    "        trace_name = [n + \"<br> (\" + str(c) + \"% coverage)\" for n,c in zip(name,AA_coverage)]\n",
    "\n",
    "        fig = go.Figure(data=new_data, layout=new_layout)\n",
    "        fig.update_layout(yaxis=dict(range=[0,len(df_plot)+1],\n",
    "                                     showticklabels=True,\n",
    "                                     tickmode = 'array',\n",
    "                                     tickvals = np.arange(0, len(df_plot))+1,\n",
    "                                     ticktext = np.array(trace_name),\n",
    "                                     showgrid=False))\n",
    "\n",
    "        y_max = len(df_plot)+1\n",
    "\n",
    "        figure_height = figure_height + (len(df_plot)*50)\n",
    "\n",
    "\n",
    "    ptm_shape_dict_sub = dict(sorted(ptm_shape_dict_sub.items()))\n",
    "    for i in range(len(ptm_shape_dict_sub)):\n",
    "        fig.add_trace(go.Scatter(y=[None],\n",
    "                                 mode='markers',\n",
    "                                 xaxis='x2',\n",
    "                                 marker=dict(symbol=list(ptm_shape_dict_sub.values())[i],\n",
    "                                             color='black'),\n",
    "                                 name=list(ptm_shape_dict_sub.keys())[i],\n",
    "                                 showlegend=True))\n",
    "\n",
    "    all_uniprot_features = list(uniprot_color_dict.keys())\n",
    "    available_features = list(set(uniprot_annotation_p_f_f.feature))\n",
    "    unique_features = [x for x in all_uniprot_features if x in available_features]\n",
    "    if len(unique_features) > 0:\n",
    "\n",
    "        y_max = y_max+1\n",
    "\n",
    "        for j in range(0,len(unique_features)):\n",
    "\n",
    "            figure_height = figure_height + 50\n",
    "\n",
    "            domain = unique_features[j]\n",
    "            domain_info_sub = uniprot_annotation_p_f_f[uniprot_annotation_p_f_f.feature==domain].reset_index(drop=True)\n",
    "            for i in range(0, domain_info_sub.shape[0]):\n",
    "                start = int(domain_info_sub.start[i])\n",
    "                end = domain_info_sub.end[i]\n",
    "                if np.isnan(domain_info_sub.end[i]):\n",
    "                    end=start #+1\n",
    "                else:\n",
    "                    end=int(end)\n",
    "\n",
    "                if domain_info_sub.feature[i] == \"STRUCTURE\":\n",
    "                    marker_col = uniprot_color_dict[domain_info_sub.annotation[i]]\n",
    "                else:\n",
    "                    marker_col = uniprot_color_dict[domain_info_sub.feature[i]]\n",
    "\n",
    "                fig.add_trace(go.Bar(x=list(range(start,end+1)),\n",
    "                                     y=list(np.repeat(0.2,end-start+1)),\n",
    "                                     base=list(np.repeat(y_max+(j/2),end-start+1)-0.1),\n",
    "                                     marker_color=marker_col,\n",
    "                                     marker_line_width=0,\n",
    "                                     opacity=0.8,\n",
    "                                     showlegend=False,\n",
    "                                     xaxis='x2',\n",
    "                                     name='',\n",
    "                                     text=np.repeat(domain_info_sub.annotation[i],len(range(start,end+1))),\n",
    "                                     hovertemplate ='%{text}'\n",
    "                                     #hovertext=domain_info_sub.annotation[i],\n",
    "                                     #hoverinfo='text'\n",
    "                                    ))\n",
    "        fig.update_layout(barmode='stack', bargap=0, hovermode='x unified',hoverdistance=1)\n",
    "    \n",
    "        y_max = y_max + (len(unique_features)/2)\n",
    "\n",
    "    selected_proteases = sorted(selected_proteases)\n",
    "    if len(selected_proteases) > 0:\n",
    "\n",
    "        y_max = y_max+1\n",
    "\n",
    "        for u in range(0,len(selected_proteases)):\n",
    "\n",
    "            figure_height = figure_height + 50\n",
    "\n",
    "            protease = selected_proteases[u]\n",
    "            sites = get_cleavage_sites(protein_sequence,protease)\n",
    "            for s in sites:\n",
    "                fig.add_trace(go.Bar(x=list(range(s+1,s+2)),\n",
    "                                     y=[0.2],\n",
    "                                     base=y_max+(u/2)-0.1,\n",
    "                                     marker_color=\"grey\",\n",
    "                                     opacity=0.8,\n",
    "                                     showlegend=False,\n",
    "                                     xaxis='x2',\n",
    "                                     name='',\n",
    "                                     text=np.repeat(protease,len(range(s+1,s+2))),\n",
    "                                     hovertemplate ='%{text}'\n",
    "                                     #hovertext=protease,\n",
    "                                     #hoverinfo='text'\n",
    "                                    ))\n",
    "                \n",
    "        y_max = y_max + (len(selected_proteases)/2)\n",
    "\n",
    "    \n",
    "    alphafold_annotation = get_alphafold_annotation(protein = protein, \n",
    "                                                    selected_features = selected_alphafold_features,\n",
    "                                                    download_folder = download_folder)\n",
    "    \n",
    "    if alphafold_annotation.empty:\n",
    "        #print('no alphafold')\n",
    "        alphafold_features = []\n",
    "    else:\n",
    "        #print('alphafold')\n",
    "        \n",
    "        alphafold_features = [i for i in alphafold_annotation.columns if i in [\n",
    "            \"AlphaFold confidence\", \n",
    "            \"AlphaFold exposure\",\n",
    "            \"AlphaFold IDR\",\n",
    "            \"AlphaFold secondary structures\"]]\n",
    "                                        \n",
    "        alphafold_color_dict = {'HELX': '#5dabe8', \n",
    "                                'STRN': '#8cdbad',\n",
    "                                'TURN': '#e094bc',\n",
    "                                'BEND': '#fac682',\n",
    "                                'unstructured': '#aaaaaa',\n",
    "                                'Very high':'#4035de',\n",
    "                                'Confident':'#6aaff7',\n",
    "                                'Low':'#f5d44e',\n",
    "                                'Very low':'#ff8f4a',\n",
    "                                'No neighbors': \"#8FF5A2\",\n",
    "                                'Few neighbors': \"#38BFA7\",\n",
    "                                'Many neighbors': \"#367BC3\",\n",
    "                                'Surrounded': \"#524582\",\n",
    "                                'IDR': \"#B5B5B5\",\n",
    "                                'folded': \"#367BC3\"}\n",
    "        \n",
    "        y_max = y_max+1\n",
    "        \n",
    "        for j in range(0,len(alphafold_features)):\n",
    "            \n",
    "            figure_height = figure_height + 50\n",
    "            \n",
    "            af_column = alphafold_features[j]\n",
    "            if af_column == \"AlphaFold confidence\":\n",
    "                extra_annotation = alphafold_annotation[\"quality\"].values\n",
    "            elif af_column == \"AlphaFold exposure\":\n",
    "                extra_annotation = alphafold_annotation[\"nAA_12_70_pae\"].values\n",
    "            else:\n",
    "                extra_annotation = np.repeat(\"\",alphafold_annotation.shape[0])\n",
    "            \n",
    "            for i in range(0, alphafold_annotation.shape[0]):\n",
    "                start = alphafold_annotation.position[i]\n",
    "                end = alphafold_annotation.position[i]  \n",
    "\n",
    "                alpha_col = alphafold_color_dict[alphafold_annotation[af_column][i]]\n",
    "\n",
    "                fig.add_trace(go.Bar(x=list(range(start,end+1)), \n",
    "                                     y=list(np.repeat(0.2,end-start+1)),\n",
    "                                     base=y_max+(j/2)-0.1,\n",
    "                                     marker_color=alpha_col,\n",
    "                                     marker_line_width=0,\n",
    "                                     opacity=0.8,\n",
    "                                     showlegend=False,\n",
    "                                     xaxis='x2',\n",
    "                                     name='',\n",
    "                                     text=np.repeat(str(alphafold_annotation[af_column][i]) + \" \" + str(extra_annotation[i]),len(range(start,end+1))),\n",
    "                                     hovertemplate ='%{text}'\n",
    "                                     #hovertext=domain_info_sub.annotation[i],\n",
    "                                     #hoverinfo='text'\n",
    "                                    ))   \n",
    "                \n",
    "        y_max = y_max + (len(alphafold_features)/2)\n",
    "    \n",
    "    fig.add_trace(go.Scatter(x=np.arange(1,len(protein_sequence)+1,1),\n",
    "                        y=np.repeat(0,len(protein_sequence)),\n",
    "                        marker=dict(color='rgba(135, 206, 250, 0)'),\n",
    "                        name='',\n",
    "                        mode='markers',\n",
    "                        xaxis='x2',\n",
    "                        text=[aa_color_dict[x] for x in list(protein_sequence)],\n",
    "                        #text=np.arange(1,len(protein_sequence)+1,1),\n",
    "                        meta=list(protein_sequence),\n",
    "                        hovertemplate ='<b>%{meta}: %{text}<b>',\n",
    "                        showlegend=False))\n",
    "\n",
    "    if figure_height < 500:\n",
    "        figure_height = 500\n",
    "\n",
    "    if dashboard:\n",
    "        plot_width = None\n",
    "    else:\n",
    "        plot_width = 1500\n",
    "\n",
    "    fig.update_layout(barmode='stack', bargap=0, hovermode='x unified',hoverdistance=1,\n",
    "                      width=plot_width, height=figure_height)\n",
    "    \n",
    "    mapped_feature_names = [uniprot_feature_dict_rev.get(key) for key in unique_features]\n",
    "    if isinstance(df, pd.DataFrame):\n",
    "        trace_ax = list([0])\n",
    "        trace_count = 1\n",
    "    elif isinstance(df, list):\n",
    "        trace_ax = list(np.arange(1,len(df_plot)+1,1))\n",
    "        trace_count = len(df_plot) + 1\n",
    "    #print(trace_ax)\n",
    "        \n",
    "    if len(unique_features) > 0:\n",
    "        trace_count += 1\n",
    "        uniprot_ax = list(np.arange(0,len(unique_features)/2, 0.5) + trace_count) \n",
    "        trace_count = max(uniprot_ax)+0.5\n",
    "        #print(uniprot_ax)\n",
    "    else:\n",
    "        uniprot_ax = []\n",
    "    if len(selected_proteases) > 0:\n",
    "        trace_count += 1\n",
    "        protease_ax = list(np.arange(0, len(selected_proteases)/2,0.5) + trace_count) \n",
    "        trace_count = max(protease_ax)+0.5\n",
    "        #print(protease_ax)\n",
    "    else:\n",
    "        protease_ax = []\n",
    "    if len(alphafold_features) > 0:\n",
    "        trace_count += 1\n",
    "        alphafold_ax = list(np.arange(0, len(alphafold_features)/2,0.5) + trace_count) \n",
    "        trace_count = max(alphafold_ax)+0.5\n",
    "        #print(alphafold_ax)\n",
    "    else:\n",
    "        alphafold_ax = []\n",
    "\n",
    "    fig.update_yaxes(showticklabels=True,\n",
    "                     #tickvals= np.arange(0, 1+len(unique_features)+len(selected_proteases)),\n",
    "                     tickvals= trace_ax + uniprot_ax + protease_ax + alphafold_ax,\n",
    "                     ticktext=np.hstack((np.array(trace_name),np.array(mapped_feature_names),np.array(selected_proteases),np.array(alphafold_features))),\n",
    "                     automargin=True,\n",
    "                     range=[-1, y_max + 0.2],\n",
    "                     showgrid=False)\n",
    "\n",
    "    #config = {'toImageButtonOptions': {'format': 'svg', # one of png, svg, jpeg, webp\n",
    "    #                                   'filename': 'custom_image',\n",
    "    #                                   'height': 500,\n",
    "    #                                   'width': 1500,\n",
    "    #                                   'scale': 1 # Multiply title/legend/axis/canvas sizes by this factor\n",
    "    #                                  }\n",
    "    #         }\n",
    "\n",
    "    return fig #.show(config=config)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#hide\n",
    "uniprot_annotation = pd.read_csv('../testdata/test_preprocessed_uniprot.csv')\n",
    "# generate mock annotations\n",
    "add_mock_annotation = pd.DataFrame(columns=uniprot_annotation.columns)\n",
    "add_mock_annotation.loc[0] = ['A0A087WTH1','HELIX',float('NaN'),10,20,'NaN']\n",
    "add_mock_annotation.loc[1] = ['A0A087WTH1','TURN',float('NaN'),25,35,'NaN']\n",
    "add_mock_annotation.loc[2] = ['A0A087WTH1','STRAND',float('NaN'),40,50,'NaN']\n",
    "# append mock annotations\n",
    "uniprot_annotation = uniprot_annotation.append(add_mock_annotation)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#hide\n",
    "all_annotatins = list(uniprot_annotation.feature.unique())\n",
    "all_annotatins.append('STRUCTURE')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#hide\n",
    "plot_peptide_traces(test_formatted_proteome_data,\n",
    "                    name = 'proteome',\n",
    "                    protein = \"A0A087WTH1\",\n",
    "                    fasta = test_fasta,\n",
    "                    uniprot=uniprot_annotation,\n",
    "                    selected_features=all_annotatins,\n",
    "                    uniprot_feature_dict=uniprot_feature_dict,\n",
    "                    uniprot_color_dict=uniprot_color_dict,\n",
    "                    selected_proteases = [\"trypsin\",'lysc'],\n",
    "                    selected_alphafold_features = [\"AlphaFold confidence\",  \n",
    "                                                   \"AlphaFold exposure\", \n",
    "                                                   \"AlphaFold IDR\", \n",
    "                                                   \"AlphaFold secondary structures\"]\n",
    "                   )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#hide\n",
    "plot_peptide_traces(test_formatted_proteome_data,\n",
    "                    name = 'proteome',\n",
    "                    protein = \"A0A087WTH1\",\n",
    "                    fasta = test_fasta,\n",
    "                    uniprot=uniprot_annotation,\n",
    "                    selected_features=all_annotatins,\n",
    "                    uniprot_feature_dict=uniprot_feature_dict,\n",
    "                    uniprot_color_dict=uniprot_color_dict,\n",
    "                    trace_colors=['red'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#hide\n",
    "plot_peptide_traces(test_formatted_proteome_data,\n",
    "                    name = 'proteome',\n",
    "                    protein = \"A0A087WTH1\",\n",
    "                    fasta = test_fasta,\n",
    "                    uniprot=uniprot_annotation,\n",
    "                    selected_features=all_annotatins,\n",
    "                    uniprot_feature_dict=uniprot_feature_dict,\n",
    "                    uniprot_color_dict=uniprot_color_dict,\n",
    "                    selected_proteases=[\"trypsin_full\",\"lysc\",\"trypsin\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#hide\n",
    "plot_peptide_traces(test_formatted_proteome_data,\n",
    "                    name = 'proteome',\n",
    "                    protein = \"A0A087WTH1\",\n",
    "                    fasta = test_fasta,\n",
    "                    uniprot=uniprot_annotation,\n",
    "                    selected_features=[],\n",
    "                    uniprot_feature_dict=uniprot_feature_dict,\n",
    "                    uniprot_color_dict=uniprot_color_dict,\n",
    "                    selected_proteases=[\"trypsin_full\",\"lysc\",\"trypsin\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#hide\n",
    "# test a modification thats not in the PTM dict\n",
    "test_formatted_proteome_data.loc[(test_formatted_proteome_data.unique_protein_id==\"A0A024R161\") & (test_formatted_proteome_data.modified_sequence==\"PEPT[Phospho (STY)]IDER\"),'modified_sequence'] = 'PE[t]PTIDER'\n",
    "test_formatted_proteome_data.loc[(test_formatted_proteome_data.unique_protein_id==\"A0A024R161\") & (test_formatted_proteome_data.modified_sequence==\"PE[t]PTIDER\"),'PTMtypes'] = ['[t]']\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#hide\n",
    "plot_peptide_traces([test_formatted_proteome_data,test_formatted_proteome_data,test_formatted_proteome_data],\n",
    "                    name = ['proteome 1','proteome 2','proteome 3'],\n",
    "                    protein = \"A0A087WTH1\",\n",
    "                    fasta = test_fasta,\n",
    "                    uniprot=uniprot_annotation,\n",
    "                    selected_features=all_annotatins,\n",
    "                    uniprot_feature_dict=uniprot_feature_dict,\n",
    "                    uniprot_color_dict=uniprot_color_dict,\n",
    "                    selected_proteases=[\"trypsin\",\"chymotrypsin high specificity\"],\n",
    "                    selected_alphafold_features = [\"AlphaFold confidence\", \n",
    "                                                   \"AlphaFold exposure\",\n",
    "                                                   \"AlphaFold IDR\",\n",
    "                                                   \"AlphaFold secondary structures\"]\n",
    "                   )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#hide\n",
    "plot_peptide_traces([test_formatted_proteome_data,test_formatted_proteome_data,test_formatted_proteome_data],\n",
    "                    name = ['proteome 1','proteome 2','proteome 3'],\n",
    "                    protein = \"A0A024R161\",\n",
    "                    fasta = test_fasta,\n",
    "                    uniprot=uniprot_annotation,\n",
    "                    selected_features=all_annotatins,\n",
    "                    uniprot_feature_dict=uniprot_feature_dict,\n",
    "                    uniprot_color_dict=uniprot_color_dict,\n",
    "                    selected_proteases=[\"trypsin\",\"chymotrypsin high specificity\"],\n",
    "                    trace_colors=[\"green\",\"red\",\"blue\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#hide\n",
    "plot_peptide_traces([test_formatted_proteome_data,test_formatted_proteome_data,test_formatted_proteome_data],\n",
    "                    name = ['proteome 1','proteome 2','proteome 3'],\n",
    "                    protein = \"A0A024R161\",\n",
    "                    fasta = test_fasta,\n",
    "                    uniprot=uniprot_annotation,\n",
    "                    selected_features=[],\n",
    "                    uniprot_feature_dict=uniprot_feature_dict,\n",
    "                    uniprot_color_dict=uniprot_color_dict,\n",
    "                    selected_proteases=[\"trypsin\",\"chymotrypsin high specificity\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#hide\n",
    "from alphamap.organisms_data import import_fasta\n",
    "human_fasta = import_fasta('Human')\n",
    "from alphamap.organisms_data import import_uniprot_annotation\n",
    "human_uniprot = import_uniprot_annotation('Human')\n",
    "from alphamap.importing import import_data\n",
    "diann_data = import_data('../testdata/test_diann_input.tsv')\n",
    "from alphamap.preprocessing import format_input_data\n",
    "formatted_diann_data = format_input_data(df=diann_data, fasta = human_fasta, modification_exp = r'\\[.*?\\]')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#hide\n",
    "plot_peptide_traces(formatted_diann_data,\n",
    "                    name = 'proteome',\n",
    "                    protein = \"Q9H7C9\",\n",
    "                    fasta = human_fasta,\n",
    "                    uniprot = human_uniprot,\n",
    "                    selected_features=['CHAIN','DOMAIN','STRUCTURE'],\n",
    "                    uniprot_feature_dict=uniprot_feature_dict, \n",
    "                    uniprot_color_dict=uniprot_color_dict,\n",
    "                    selected_alphafold_features = [\"AlphaFold confidence\", \n",
    "                                                   \"AlphaFold exposure\",\n",
    "                                                   \"AlphaFold IDR\",\n",
    "                                                   \"AlphaFold secondary structures\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3D visualization"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#export \n",
    "def extract_annotation(df: pd.DataFrame):\n",
    "    df = df[df.marker_symbol>0]\n",
    "    df = df.astype({\"PTMtype\": str})\n",
    "    df['MS data'] = ['detected' if m=='nan' else m for m in df.PTMtype]\n",
    "    df = df[['seq_position','MS data']]\n",
    "    df['MS data'] = [x.replace(\" \", \"\") for x in df['MS data']]\n",
    "    df = df.rename(columns={\"seq_position\": \"position\"})\n",
    "    df['position'] = df['position']+1\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#export \n",
    "def manipulate_cif(protein: str, \n",
    "                   MS_data: pd.DataFrame,\n",
    "                   download_folder: str = tempfile.gettempdir(),\n",
    "                   ):\n",
    "    \n",
    "    \n",
    "    #print(MS_data)\n",
    "    \n",
    "    alphafold_annotation = get_alphafold_annotation(protein = protein, \n",
    "                                                    selected_features = [\"AlphaFold confidence\", \n",
    "                                                                         \"AlphaFold exposure\",\n",
    "                                                                         \"AlphaFold IDR\",\n",
    "                                                                         \"AlphaFold secondary structures\"],\n",
    "                                                    download_folder = download_folder)\n",
    "    \n",
    "    if alphafold_annotation.empty:\n",
    "        return False\n",
    "    \n",
    "    else:\n",
    "        mod_table = False\n",
    "        AA_position = 0\n",
    "        infile = os.path.join(download_folder, protein+'.cif')\n",
    "        outfile = os.path.join(download_folder, protein+'_mod.cif')\n",
    "\n",
    "        with open(infile) as f:\n",
    "            lines = []\n",
    "            for line in f:\n",
    "                line = line.rstrip()\n",
    "                if mod_table:\n",
    "                    if line == '#':\n",
    "                        mod_table = False\n",
    "                        lines.append(line)\n",
    "                    else:\n",
    "                        AA_position += 1\n",
    "\n",
    "                        # append exposure\n",
    "                        exposure = str(alphafold_annotation[alphafold_annotation.position==AA_position].nAA_12_70_pae.values[0])\n",
    "                        n_extra_blank = 4-len(exposure)\n",
    "                        appended_line = line+' '+exposure+(n_extra_blank*' ')\n",
    "                        # append MS data\n",
    "                        if AA_position in list(MS_data.position.values):\n",
    "                            appended_line = appended_line+' '+'detected'+' '\n",
    "                            ms_evidence = MS_data[MS_data.position==AA_position]['MS data'].values[0]\n",
    "                            n_extra_blank = 40-len(ms_evidence)\n",
    "                            appended_line = appended_line+' '+ms_evidence+(n_extra_blank*' ')\n",
    "                        else:\n",
    "                            appended_line = appended_line+' '+'absent  '+' '\n",
    "                            n_extra_blank = 40-len('absent')\n",
    "                            appended_line = appended_line+' '+'absent'+(n_extra_blank*' ')\n",
    "                        # append PTM data\n",
    "                        ptm_sub = MS_data[MS_data['MS data'] != 'detected']\n",
    "                        if AA_position in list(ptm_sub.position.values):\n",
    "                            ptm_evidence = ptm_sub[ptm_sub.position==AA_position]['MS data'].values[0]\n",
    "                            n_extra_blank = 40-len(ptm_evidence)\n",
    "                            appended_line = appended_line+' '+ptm_evidence+(n_extra_blank*' ')\n",
    "                        else:\n",
    "                            n_extra_blank = 40-len('none')\n",
    "                            appended_line = appended_line+' '+'none'+(n_extra_blank*' ')\n",
    "                        # append IDR\n",
    "                        IDR_info = alphafold_annotation[alphafold_annotation.position==AA_position]['AlphaFold IDR'].values[0]\n",
    "                        n_extra_blank = 10-len(IDR_info)\n",
    "                        appended_line = appended_line+' '+IDR_info+(n_extra_blank*' ')\n",
    "                        # append IDR + PTM data\n",
    "                        if AA_position in list(ptm_sub.position.values):\n",
    "                            n_extra_blank = 40-len(ptm_evidence)\n",
    "                            appended_line = appended_line+' '+ptm_evidence+(n_extra_blank*' ')\n",
    "                        else:\n",
    "                            n_extra_blank = 40-len(IDR_info)\n",
    "                            appended_line = appended_line+' '+IDR_info+(n_extra_blank*' ')\n",
    "                        # append structure group\n",
    "                        structure_info = alphafold_annotation[alphafold_annotation.position==AA_position]['AlphaFold secondary structures'].values[0]\n",
    "                        n_extra_blank = 15-len(structure_info)\n",
    "                        appended_line = appended_line+' '+structure_info+(n_extra_blank*' ')\n",
    "\n",
    "                        lines.append(appended_line)\n",
    "                else:\n",
    "                    if line == '_ma_qa_metric_local.ordinal_id':\n",
    "                        lines.append(line)\n",
    "                        lines.append('_ma_qa_metric_local.exposure')\n",
    "                        lines.append('_ma_qa_metric_local.aa')\n",
    "                        lines.append('_ma_qa_metric_local.aa_ptm')\n",
    "                        lines.append('_ma_qa_metric_local.ptm')\n",
    "                        lines.append('_ma_qa_metric_local.idr')\n",
    "                        lines.append('_ma_qa_metric_local.idr_ptm')\n",
    "                        lines.append('_ma_qa_metric_local.structure_group')\n",
    "                        mod_table = True\n",
    "                    else:\n",
    "                        lines.append(line)\n",
    "\n",
    "        try:\n",
    "            os.remove(outfile)\n",
    "        except:\n",
    "            pass\n",
    "\n",
    "        with open(outfile, 'a') as outfile:\n",
    "            for l in lines:\n",
    "                outfile.write(l+'\\n')\n",
    "                \n",
    "        return True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#export \n",
    "def adjust_html(protein: str, \n",
    "                coloring: str, \n",
    "                data_name: str):\n",
    "    \n",
    "    html_template = \"\"\"\n",
    "                    <!DOCTYPE html>\n",
    "                    <html lang=\"en\">\n",
    "                      <head>\n",
    "                        <meta charset=\"utf-8\" />\n",
    "                        <meta name=\"viewport\" content=\"width=device-width, user-scalable=no, minimum-scale=1.0, maximum-scale=1.0\">\n",
    "                        <title>PDBe Molstar</title>\n",
    "\n",
    "                        <!-- Molstar CSS & JS -->\n",
    "                        <link rel=\"stylesheet\" type=\"text/css\" href=\"js/alphafold-viz.css\">\n",
    "                        <script type=\"text/javascript\" src=\"js/alphafold-viz.js\"></script>\n",
    "                        <style>\n",
    "                          * {\n",
    "                              margin: 0;\n",
    "                              padding: 0;\n",
    "                              box-sizing: border-box;\n",
    "                          }\n",
    "                          .msp-plugin ::-webkit-scrollbar-thumb {\n",
    "                              background-color: #5b5b68 !important;\n",
    "                          }\n",
    "                          .viewerSection {\n",
    "                            padding-top: 100px;\n",
    "                          }\n",
    "                          .controlsSection {\n",
    "                            width: 300px;\n",
    "                            float: left;\n",
    "                            padding: 40px 0 0 40px;\n",
    "                            margin-right: 30px;\n",
    "                          }\n",
    "                          .controlBox {\n",
    "                            border: 1px solid lightgray;\n",
    "                            padding: 10px;\n",
    "                            margin-bottom: 20px;\n",
    "                          }\n",
    "                          #myViewer{\n",
    "                            float:left;\n",
    "                            width: 800px;\n",
    "                            height: 800px;\n",
    "                            position:relative;\n",
    "                          }\n",
    "                        </style>\n",
    "                      </head>\n",
    "\n",
    "                      <body>\n",
    "\n",
    "\n",
    "\n",
    "                        <div class=\"viewerSection\">\n",
    "                          <!-- <h4>3D viewer</h4> -->\n",
    "                          <h4>   </h4>\n",
    "\n",
    "                          <!-- Molstar container -->\n",
    "                          <div id=\"myViewer\"></div>\n",
    "\n",
    "                        </div>\n",
    "                        <script>\n",
    "\n",
    "                          globalThis.color_choice = your_choice //'af', 'access', 'ptm'\n",
    "\n",
    "                          //Create plugin instance\n",
    "                          var viewerInstance = new PDBeMolstarPlugin();\n",
    "\n",
    "                          //Set options (Checkout available options list in the documentation)\n",
    "                          var options = {\n",
    "                            customData: {\n",
    "                              url: 'cif/your_file.cif',\n",
    "                              format: 'cif'\n",
    "                            },\n",
    "                            hideControls: false,\n",
    "                            // expanded: true,\n",
    "                            isAfView: true,\n",
    "                            hideCanvasControls: [] //['selection', 'animation', 'controlToggle', 'controlInfo']\n",
    "                          }\n",
    "\n",
    "                          //Get element from HTML/Template to place the viewer \n",
    "                          var viewerContainer = document.getElementById('myViewer');\n",
    "\n",
    "                          //Call render method to display the 3D view\n",
    "                          viewerInstance.render(viewerContainer, options);\n",
    "                          viewerInstance.canvas.setBgColor({r:255, g:255, b:255});\n",
    "\n",
    "                        </script>\n",
    "                      </body>\n",
    "\n",
    "                    </html>\n",
    "                    \"\"\"\n",
    "    \n",
    "    html_template = html_template.replace('your_file.cif',protein+'_'+data_name+'_mod.cif')\n",
    "    \n",
    "    if coloring == 'AlphaFold confidence':\n",
    "        color = 'af'\n",
    "    elif coloring == 'AlphaFold exposure':\n",
    "        color = 'exposure'\n",
    "    elif coloring == 'MS peptides':\n",
    "        color = 'aa'\n",
    "    elif coloring == 'MS modified peptides':\n",
    "        color = 'aa_ptm'\n",
    "    elif coloring == 'MS PTMs':\n",
    "        color = 'ptm'\n",
    "    elif coloring == 'AlphaFold IDR':\n",
    "        color = 'idr'\n",
    "    elif coloring == 'MS PTMs on AlphaFold IDR':\n",
    "        color = 'idr_ptm'\n",
    "    elif coloring == 'AlphaFold secondary structures':\n",
    "        color = 'structure_group'\n",
    "    else:\n",
    "        color = 'af'\n",
    "        \n",
    "    html = html_template.replace('globalThis.color_choice = your_choice',f\"globalThis.color_choice = '{color}'\")\n",
    "     \n",
    "    return html "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#export\n",
    "def get_ms_concensus(ms_list):\n",
    "    unique_ms_list = set(ms_list)\n",
    "    if len(unique_ms_list)==1:\n",
    "        out = unique_ms_list\n",
    "    else:\n",
    "        out = [i for i in unique_ms_list if i != 'detected']\n",
    "\n",
    "    out = ';'.join(out)\n",
    "    \n",
    "    return out"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#export \n",
    "import alphamap\n",
    "def plot_3d_structure(df: pd.DataFrame or list,\n",
    "                        name: str,\n",
    "                        protein: str,\n",
    "                        fasta: fasta,\n",
    "                        selected_coloring: str,\n",
    "                        dashboard: bool = False, \n",
    "                        download_folder: str = tempfile.gettempdir()):\n",
    "\n",
    "    \"\"\"\n",
    "    Function to generate the 3D sequence plot.\n",
    "\n",
    "    Args:\n",
    "        df (pd.DataFrame/list): Single dataframe or list of dataframes containing the datasets to plot.\n",
    "        name (str): Single string containing the name of the MS dataset in df.\n",
    "        protein (str): Uniprot protein accession.\n",
    "        fasta (fasta): Fasta file imported by pyteomics 'fasta.IndexedUniProt'.\n",
    "        selected_coloring (str): Coloring to show.\n",
    "        dashboard (bool, optional): Flag if the function is called from the dashboard. Default is 'False'.\n",
    "        \n",
    "    Returns:\n",
    "        go.Figure: 3D plot.\n",
    "\n",
    "    \"\"\"\n",
    "\n",
    "    if isinstance(df, pd.DataFrame):\n",
    "        df_plot = get_plot_data(protein=protein,\n",
    "                                df = df,\n",
    "                                fasta = fasta)\n",
    "    elif isinstance(df, list):\n",
    "        df_plot = [get_plot_data(protein=protein,\n",
    "                                 df = d,\n",
    "                                 fasta = fasta) for d in df]\n",
    "        df_plot = pd.concat(df_plot)\n",
    "        \n",
    "    if df_plot is None:\n",
    "        MS_annotation =  pd.DataFrame({'position' : [], 'MS data': []})   \n",
    "    else:\n",
    "        MS_annotation = extract_annotation(df_plot) \n",
    "        MS_annotation = MS_annotation.groupby('position')['MS data'].apply(list).reset_index(name='MS data')\n",
    "        MS_annotation['MS data'] = MS_annotation['MS data'].apply(lambda x: get_ms_concensus(x))\n",
    "\n",
    "    cif_available = manipulate_cif(protein = protein, \n",
    "                                   MS_data = MS_annotation,\n",
    "                                   download_folder = download_folder)\n",
    "    \n",
    "    BASE_PATH = os.path.dirname(alphamap.__file__)\n",
    "    js_path = os.path.join(BASE_PATH, 'js')\n",
    "    cif_path = download_folder\n",
    "    \n",
    "    \n",
    "    import shutil\n",
    "    new_file = os.path.join(cif_path, protein+'_'+name+'_mod.cif')\n",
    "    try:\n",
    "        os.remove(new_file)\n",
    "    except:\n",
    "        pass\n",
    "    \n",
    "\n",
    "    if cif_available:\n",
    "        shutil.copyfile(\n",
    "            src = os.path.join(cif_path, protein+'_mod.cif'),\n",
    "            dst = new_file)\n",
    "        \n",
    "        mod_html = adjust_html(protein=protein, \n",
    "                               coloring=selected_coloring,\n",
    "                               data_name=name)\n",
    "    else:\n",
    "        mod_html = f\"\"\"\n",
    "            <!DOCTYPE html>\n",
    "            <center> No AlphaFold structure available for {protein}. <center>\n",
    "            \"\"\"\n",
    "    \n",
    "    return mod_html, js_path, cif_path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#export \n",
    "def format_for_3Dviz(\n",
    "    df: pd.DataFrame,\n",
    "    ptm_dataset: str\n",
    ") -> pd.DataFrame:\n",
    "    \"\"\"\n",
    "    Function to format data for 3D visualization.\n",
    "\n",
    "    Args:\n",
    "        df (pd.DataFrame): Single dataframe containing PTM data for visualization, \n",
    "        formatted accorinding to StructureMap.\n",
    "        ptm_dataset (str): Single string containing the name of the target PTM column in df.\n",
    "        \n",
    "    Returns:\n",
    "        pd.DataFrame: DataFrame containing the formatted PTM dat\n",
    "\n",
    "    \"\"\"\n",
    "    df_mod = df[[\"protein_id\",\"AA\",\"position\",ptm_dataset]]\n",
    "    df_mod = df_mod.rename(columns={\"protein_id\": \"unique_protein_id\",\n",
    "                                    \"AA\": \"modified_sequence\",\n",
    "                                    \"position\": \"start\"})\n",
    "    df_mod[\"modified_sequence\"] = [mod+\"_\"+str(i) for i,mod in enumerate(df_mod[\"modified_sequence\"])]\n",
    "    df_mod[\"all_protein_ids\"] = df_mod[\"unique_protein_id\"]\n",
    "    df_mod[\"PTMsites\"] = 0\n",
    "    df_mod[\"start\"] = df_mod[\"start\"]-1\n",
    "    df_mod[\"end\"] = df_mod[\"start\"]\n",
    "    df_mod[\"PTMsites\"] = [[i] for i in df_mod[\"PTMsites\"]]\n",
    "    df_mod = df_mod[df_mod[ptm_dataset] == 1]\n",
    "    df_mod[\"marker_symbol\"] = 1\n",
    "    df_mod[\"PTMtypes\"] = [[ptm_dataset] for i in df_mod[\"PTMsites\"]]\n",
    "    df_mod = df_mod.dropna(subset=['PTMtypes']).reset_index(drop=True)\n",
    "    return df_mod"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#export\n",
    "import os\n",
    "import tempfile\n",
    "import panel as pn\n",
    "pn.extension()\n",
    "def visualize_structure_in_panel(\n",
    "    plot_html: str,\n",
    "    js_path: str,\n",
    "    cif_path: str\n",
    "):\n",
    "    BASE_PATH = os.path.dirname(alphamap.__file__)\n",
    "    js_path_global = os.path.join(BASE_PATH, 'js')\n",
    "    cif_path_global = tempfile.gettempdir()\n",
    "    pn.extension(js_files={'js': 'js/alphafold-viz.js'})\n",
    "    pn.extension(css_files=['js/alphafold-viz.css'])\n",
    "    app = pn.Row(plot_html)\n",
    "    plot = pn.Row(pn.Column(\n",
    "        pn.layout.HSpacer(height=80),\n",
    "        pn.panel(plot_html, width = 1500, align='center', sizing_mode='stretch_width'),\n",
    "        align='center', sizing_mode='stretch_width'),\n",
    "                  align='center', sizing_mode='stretch_width')\n",
    "    pn.serve(plot, static_dirs={'js': js_path, 'cif': cif_path}, verbose=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# #hide\n",
    "# mod_html, js_path, cif_path = plot_3d_structure(df = formatted_diann_data,\n",
    "#                     name = 'proteome',\n",
    "#                     protein = \"P23284\", #Q9H7C9, P23284 ## not working example: Q15149\n",
    "#                     fasta = human_fasta,\n",
    "#                     selected_coloring = 'AlphaFold quality', #'AlphaFold exposure', #'AlphaFold quality', 'MS data'\n",
    "#                     dashboard = False)\n",
    "\n",
    "# visualize_structure_in_panel(mod_html, js_path, cif_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# options: \n",
    "# 'alphafold_quality'\n",
    "# 'alphafold_exposure'\n",
    "# 'alphafold_IDR'\n",
    "# 'alphafold_secondary_structures'\n",
    "# 'MS data'\n",
    "# 'MS PTMs'\n",
    "# 'MS amino acids'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#export \n",
    "import os\n",
    "import tempfile\n",
    "from alphamap.organisms_data import import_fasta\n",
    "def plot_3d_structuremap(\n",
    "    df: pd.DataFrame,\n",
    "    organism: str,\n",
    "    protein: str,\n",
    "    ptm_type: str\n",
    "):\n",
    "    \"\"\"\n",
    "    Function to return a sigle \n",
    "\n",
    "    Args:\n",
    "        df (pd.DataFrame): Single dataframe containing PTM data for visualization, \n",
    "        organism (str): String specifying the organism.\n",
    "        protein (str): String specufying the UniProt protein accession.\n",
    "        ptm_type (str): String of the PTM type to visualize. \n",
    "        \n",
    "    Returns:\n",
    "        go.Figure: 3D plot.\n",
    "\n",
    "    \"\"\"\n",
    "    fasta = import_fasta(organism)\n",
    "    df_sub = df[df.protein_id==protein]\n",
    "    formatted_data = format_for_3Dviz(\n",
    "        df=df_sub, ptm_dataset=ptm_type) \n",
    "    plot3D_html, js_path, cif_path = plot_3d_structure(df = formatted_data,\n",
    "                    name = 'proteome',\n",
    "                    protein = protein,\n",
    "                    fasta = fasta,\n",
    "                    selected_coloring = 'MS PTMs', \n",
    "                    dashboard = False)\n",
    "    return plot3D_html, js_path, cif_path"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Create a pdf report"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#export\n",
    "from alphamap.pdflib import *\n",
    "\n",
    "def create_pdf_report(proteins: list,\n",
    "                      df: pd.DataFrame or list,\n",
    "                      name: str or list,\n",
    "                      fasta: fasta,\n",
    "                      uniprot: pd.DataFrame,\n",
    "                      selected_features: list,\n",
    "                      uniprot_feature_dict: dict,\n",
    "                      uniprot_color_dict: dict,\n",
    "                      selected_proteases: list = [],\n",
    "                      trace_colors: list = []):\n",
    "    \"\"\"\n",
    "    Function to write pdf reports for selected proteins\n",
    "\n",
    "    Args:\n",
    "        proteins (list): List of uniprot protein accessions.\n",
    "        df (pd.DataFrame/list): Single dataframe or list of dataframes containing the datasets to plot.\n",
    "        name (str/list): Single string or list of strings containing the names for each dataset in df.\n",
    "        fasta (fasta): Fasta file imported by pyteomics 'fasta.IndexedUniProt'.\n",
    "        uniprot (pd.DataFrame): Uniprot annotations formatted by alphamap.\n",
    "        selected_features (list): List of uniprot features to plot.\n",
    "        uniprot_feature_dict (dict): Uniprot feature dictionary.\n",
    "        uniprot_color_dict (dict): Uniprot color dictionary.\n",
    "        selected_proteases (list, optional): List of proteases to plot. Default is an empty list.\n",
    "        trace_colors (list, optional): List of manualy selected colors for each dataset in df. Default is an empty list.\n",
    "\n",
    "    Returns:\n",
    "        BytesIO: BytesIO object for writing a pdf report.\n",
    "    \"\"\"\n",
    "\n",
    "    if isinstance(df, pd.DataFrame):\n",
    "        max_height = 200 + 50 + (len(selected_features)*50) + (len(selected_proteases)*50)\n",
    "    else:\n",
    "        max_height = 200 + (len(df)*50) + (len(selected_features)*50) + (len(selected_proteases)*50)\n",
    "\n",
    "\n",
    "    if max_height < 700:\n",
    "        max_height = 700\n",
    "\n",
    "    footer_text = '<font size=\"20\">This report was generated by <a href=\"https://github.com/MannLabs/alphamap\" color=\"darkblue\"><b>AlphaMap</b></a>.</font>'\n",
    "\n",
    "    pdf_buf = BytesIO()\n",
    "    pdf_report = canvas.Canvas(pdf_buf, pagesize=(1600,max_height))\n",
    "\n",
    "    if isinstance(proteins, list):\n",
    "        for p in range(0,len(proteins)):\n",
    "            plot = plot_peptide_traces(df=df, name=name, protein=proteins[p], fasta=fasta,\n",
    "                                       uniprot=uniprot, selected_features=selected_features,\n",
    "                                       uniprot_feature_dict=uniprot_feature_dict,\n",
    "                                       uniprot_color_dict=uniprot_color_dict,\n",
    "                                       selected_proteases=selected_proteases,\n",
    "                                       trace_colors=trace_colors)\n",
    "            draw_content(pdf_report, plot, width=1600, height=max_height,\n",
    "                         spacing=5, border=20)\n",
    "            draw_content(pdf_report, footer_text, width=1600, height=100,\n",
    "                         spacing=20, border=30)\n",
    "            pdf_report.showPage()\n",
    "    else:\n",
    "        plot = plot_peptide_traces(df=df, name=name, protein=proteins, fasta=fasta,\n",
    "                                   uniprot=uniprot, selected_features=selected_features,\n",
    "                                   uniprot_feature_dict=uniprot_feature_dict,\n",
    "                                   uniprot_color_dict=uniprot_color_dict,\n",
    "                                   selected_proteases=selected_proteases,\n",
    "                                   trace_colors=trace_colors)\n",
    "        draw_content(pdf_report, plot, width=1600, height=max_height,\n",
    "                         spacing=20, border=30)\n",
    "        draw_content(pdf_report, footer_text, width=1600, height=100,\n",
    "                         spacing=20, border=30)\n",
    "        pdf_report.showPage()\n",
    "\n",
    "    pdf_report.save()\n",
    "    pdf_buf.seek(0)\n",
    "\n",
    "    #with open(\"file.pdf\", \"wb\") as file:\n",
    "    #    file.write(pdf_buf.getvalue())\n",
    "\n",
    "    return pdf_buf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#hide\n",
    "p = create_pdf_report(proteins = [\"A0A024R161\",\"A0A087WT10\",\"A0A087WTH1\",\"A0A087WTH5\"],\n",
    "                      df=[test_formatted_proteome_data,test_formatted_proteome_data,test_formatted_proteome_data],\n",
    "                    name = ['proteome 1','proteome 2','proteome 3'],\n",
    "                    fasta = test_fasta,\n",
    "                    uniprot=uniprot_annotation,\n",
    "                    selected_features=all_annotatins,\n",
    "                    uniprot_feature_dict=uniprot_feature_dict,\n",
    "                    uniprot_color_dict=uniprot_color_dict,\n",
    "                    selected_proteases=[\"trypsin\",\"chymotrypsin high specificity\",\n",
    "                                        \"caspase 1\",\"caspase 2\",\"caspase 3\"])\n",
    "\n",
    "#with open(\"file.pdf\", \"wb\") as file:\n",
    "#    file.write(p.getvalue())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#hide\n",
    "p = create_pdf_report(proteins = [\"A0A024R161\", \"A0A087WT10\"],\n",
    "                      df=test_formatted_proteome_data,\n",
    "                    name = 'proteome 1',\n",
    "                    fasta = test_fasta,\n",
    "                    uniprot=uniprot_annotation,\n",
    "                    selected_features=all_annotatins,\n",
    "                    uniprot_feature_dict=uniprot_feature_dict,\n",
    "                    uniprot_color_dict=uniprot_color_dict,\n",
    "                    selected_proteases=[\"trypsin\"])\n",
    "\n",
    "#with open(\"file.pdf\", \"wb\") as file:\n",
    "#    file.write(p.getvalue())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#hide\n",
    "\n",
    "###### Export notebook to script ###### "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#hide\n",
    "from nbdev.showdoc import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#hide\n",
    "from nbdev.export import *\n",
    "notebook2script()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:alphamap_fold]",
   "language": "python",
   "name": "conda-env-alphamap_fold-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
