---

title: AlphaMap


keywords: fastai
sidebar: home_sidebar

summary: "A python-based library that enables the exploration of proteomic datasets on the peptide level."
description: "A python-based library that enables the exploration of proteomic datasets on the peptide level."
nb_path: "nbs/index.ipynb"
---
<!--

#################################################
### THIS FILE WAS AUTOGENERATED! DO NOT EDIT! ###
#################################################
# file to edit: nbs/index.ipynb
# command to build the docs after a change: nbdev_build_docs

-->

<div class="container" id="notebook-container">
        
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="About">About<a class="anchor-link" href="#About"> </a></h2><p>AlphaMap is a tool for peptide level MS data exploration. You can load and inspect MS data analyzed by <a href="https://github.com/MannLabs/alphapept">AlphaPept</a>, DIA-NN, MaxQuant or Spectronaut. Uploaded data is processed and formatted for visual inspection of the sequence coverage of any selected protein and its identified post-translational modifications (PTMs). UniProt information is available to directly annotate sequence regions of interest such as protein domains, secondary structures, sequence variants, known PTMs, etc. Additionally, users can select proteases to further evaluate the distribution of proteolytic cleavage sites across a protein sequence. The functionality of AlphaMap can be accessed via an intuitive graphical user interface or – more flexibly – as a Python package that allows its integration into common analysis workflows for data visualization.</p>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="License">License<a class="anchor-link" href="#License"> </a></h2><p>AlphaMap was developed by the <a href="https://www.biochem.mpg.de/mann">Mann Labs at the Max Planck Institute of Biochemistry</a> and is freely available with an <a href="LICENSE">Apache License</a>.</p>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Installation">Installation<a class="anchor-link" href="#Installation"> </a></h2><p>AlphaMap can be installed and used on Windows and MacOS.
There are three different types of installation possible:</p>
<ul>
<li><a href="#one-click-gui"><strong>One-click GUI installer:</strong></a> Choose this installation if you only want the GUI and/or keep things as simple as possible.</li>
<li><a href="#pip"><strong>Pip installer:</strong></a> Choose this installation if you want to use AlphaMap as a Python package in an existing Python 3.8 environment (e.g. a Jupyter notebook). If needed, the GUI can be installed with pip as well.</li>
<li><a href="#developer"><strong>Developer installer:</strong></a> Choose this installation if you are familiar with CLI tools, <a href="https://docs.conda.io/en/latest/">conda</a> and Python. This installation allows access to all available features of AlphaMap and even allows to modify its source code directly.</li>
</ul>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h3 id="One-click-GUI">One-click GUI<a class="anchor-link" href="#One-click-GUI"> </a></h3><p>The GUI of AlphaMap is a completely stand-alone tool that requires no knowledge of Python. Click on one of the links below to download the latest release for:</p>
<ul>
<li><a href="https://github.com/MannLabs/alphamap/releases/latest/download/alphamap_installer_windows.exe"><strong>Windows</strong></a></li>
<li><a href="https://github.com/MannLabs/alphamap/releases/latest/download/alphamap_gui_installer_macos.pkg"><strong>MacOS</strong></a></li>
</ul>
<p><strong><em>IMPORTANT: Please refer to the <a href="alphamap/data/alphamap_tutorial.pdf">GUI manual</a> for detailed instructions on the installation, troubleshooting and usage of the stand-alone AlphaMap GUI.</em></strong></p>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h3 id="Pip">Pip<a class="anchor-link" href="#Pip"> </a></h3><p>AlphaMap can be installed in an existing Python 3.8 environment with a single <code>bash</code> command. <em>This <code>bash</code> command can also be run directly from within a Jupyter notebook by prepending it with a <code>!</code></em>.</p>
<div class="highlight"><pre><span></span>pip install alphamap
</pre></div>
<p>When a new version of AlphaMap becomes available, the old version can easily be upgraded by running e.g. the command again with an additional <code>--upgrade</code> flag:</p>
<div class="highlight"><pre><span></span>pip install alphamap --upgrade
</pre></div>
<p>NOTE: When installing with <code>pip</code>, UniProt information is not included. Upon first usage of a specific Organism, its information will be automatically downloaded from UniProt.</p>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h3 id="Developer">Developer<a class="anchor-link" href="#Developer"> </a></h3><p>AlphaMap can also be installed in editable (i.e. developer) mode with a few <code>bash</code> commands. This allows to fully customize the software and even modify the source code to your specific needs. When an editable Python package is installed, its source code is stored in a transparent location of your choice. While optional, it is advised to first (create and) navigate to e.g. a general software folder:</p>
<div class="highlight"><pre><span></span>mkdir ~/folder/where/to/install/software
<span class="nb">cd</span> ~/folder/where/to/install/software
</pre></div>
<p>Next, download the AlphaMap repository from GitHub either directly or with a <code>git</code> command. This creates a new AlphaMap subfolder in your current directory.</p>
<div class="highlight"><pre><span></span>git clone https://github.com/MannLabs/alphamap.git
<span class="nb">cd</span> alphamap
</pre></div>
<p>For any Python package, it is highly recommended to use a <a href="https://docs.conda.io/en/latest/">conda virtual environment</a>. AlphaMap can either be installed in a new conda environment or in an already existing environment. <em>Note that dependency conflicts can occur with already existing packages in the latter case</em>! Once a conda environment is activated, AlphaMap and all its <a href="requirements">dependencies</a> need to be installed.</p>
<div class="highlight"><pre><span></span>conda create -n alphamap <span class="nv">python</span><span class="o">=</span><span class="m">3</span>.8 -y
conda activate alphamap
pip install -e .
</pre></div>
<ul>
<li><p>By using the editable flag <code>-e</code>, all modifications to the AlphaMap <a href="alphamap">source code folder</a> are directly reflected when running AlphaMap. Note that the AlphaMap folder cannot be moved and/or renamed if an editable version is installed.</p>
</li>
<li><p>When using Jupyter notebooks and multiple conda environments direcly from the terminal, it is recommended to <code>conda install nb_conda_kernels</code> in the conda base environment. Hereafter, running a <code>jupyter notebook</code> from the conda base environment should have a <code>python [conda env: alphamap]</code> kernel available, in addition to all other conda kernels in which the command <code>conda install ipykernel</code> was run.</p>
</li>
</ul>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Test-data">Test data<a class="anchor-link" href="#Test-data"> </a></h2><p>AlphaMap has direct data import options for MaxQuant and Spectronaut.</p>
<h3 id="AlphaPept">AlphaPept<a class="anchor-link" href="#AlphaPept"> </a></h3><p>AlphaMap takes the <em>results.csv</em> file from AlphaPept as input format. An example is available for <a href="https://github.com/MannLabs/alphamap/releases/download/v0.0.210730-alpha/test_alphapept_input.csv">download here</a>.</p>
<h3 id="DIA-NN">DIA-NN<a class="anchor-link" href="#DIA-NN"> </a></h3><p>AlphaMap takes the peptide-level output .tsv file from DIA-NN as input format. An example is available for <a href="https://github.com/MannLabs/alphamap/releases/download/v0.0.210730-alpha/test_diann_input.tsv">download here</a>.</p>
<h3 id="MaxQuant">MaxQuant<a class="anchor-link" href="#MaxQuant"> </a></h3><p>AlphaMap takes the <em>evidence.txt</em> file from MaxQuant as input format. A reduced example file is available for <a href="https://github.com/MannLabs/alphamap/releases/download/v0.0.210622-alpha/test_maxquant_input.txt">download here</a>.</p>
<h3 id="Spectronaut">Spectronaut<a class="anchor-link" href="#Spectronaut"> </a></h3><p>AlphaMap takes Spectronaut results exported in normal long format (.csv or .tsv) as input. Necessary columns include:</p>
<ul>
<li>PEP.AllOccuringProteinAccessions</li>
<li>EG.ModifiedSequence</li>
<li>R.FileName</li>
</ul>
<p>To ensure proper formatting of the Spectronaut output, an export scheme is available for <a href="https://github.com/MannLabs/alphamap/blob/master/alphamap/data/spectronaut_export_scheme.rs">download here</a>.</p>
<p>A reduced example file is also available for <a href="https://github.com/MannLabs/alphamap/releases/download/v0.0.210622-alpha/test_spectronaut_input.tsv">download here</a>.
<!-- It is not directly clear how to download this individual file from here. Luckily, the two larger ones have a "download" button on the top right --></p>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Usage">Usage<a class="anchor-link" href="#Usage"> </a></h2><p>There are two ways to use AlphaMap:</p>
<ul>
<li><a href="#gui"><strong>GUI:</strong></a> This allows to interactively import and visualize the data.</li>
<li><a href="#python-and-jupyter-notebooks"><strong>Python:</strong></a> This allows to access data and explore it interactively with custom code.</li>
</ul>
<p>NOTE: The first time you use a fresh installation of AlphaMap, it is often quite slow because some functions might still need compilation on your local operating system and architecture. Subsequent use should be a lot faster.</p>
<h3 id="GUI">GUI<a class="anchor-link" href="#GUI"> </a></h3><p>Please refer to the <a href="https://github.com/MannLabs/alphamap/blob/master/alphamap/data/alphamap_tutorial.pdf">GUI manual</a> for detailed instructions on the installation and usage of the stand-alone AlphaMap GUI.</p>
<p>If the GUI was not installed through a one-click GUI installer, it can be activated with the following <code>bash</code> command:</p>
<div class="highlight"><pre><span></span>alphamap
</pre></div>
<p>Note that this needs to be prepended with a <code>!</code> when you want to run this from within a Jupyter notebook. When the command is run directly from the command-line, make sure you use the right environment (activate it with e.g. <code>conda activate alphamap</code> or set an alias to the binary executable).</p>
<h3 id="Python-and-Jupyter-notebooks">Python and Jupyter notebooks<a class="anchor-link" href="#Python-and-Jupyter-notebooks"> </a></h3><p>AlphaMap can be imported as a Python package into any Python script or notebook with the command <code>import alphamap</code>.</p>
<p>A Jupyter notebook tutorial <a href="Workflow.ipynb">'Workflow.ipynb'</a> is available to demonstrate how to load AlphaMap as python module and hot to visualize data interactively. When running locally it provides interactive plots, which are not rendered on GitHub.</p>
<p>AlphaMap includes fasta files and UniProt annotations for: 'Human', 'Mouse', 'Rat', 'Cow', 'Zebrafish', 'Drosophila', 'Caenorhabditis elegans', 'Slime mold', 'Arabidopsis thaliana', 'Rice', 'Escherichia coli', 'Bacillus subtilis', 'Saccharomyces cerevisiae', 'SARS-COV' and 'SARS-COV-2'. If additional organisms are of interest, corresponding .fasta files and sequence annotations can be downloaded directly from UniProt. A Jupyter notebook tutorial <a href="Uniprot_preprocessing.ipynb">'Uniprot_preprocessing.ipynb'</a> shows how to load and format a UniProt annotation file.</p>

</div>
</div>
</div>
</div>
 

