---

title: Protease information


keywords: fastai
sidebar: home_sidebar



nb_path: "nbs/proteolytic_cleavage.ipynb"
---
<!--

#################################################
### THIS FILE WAS AUTOGENERATED! DO NOT EDIT! ###
#################################################
# file to edit: nbs/proteolytic_cleavage.ipynb
# command to build the docs after a change: nbdev_build_docs

-->

<div class="container" id="notebook-container">
        
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Protease-dictionary">Protease dictionary<a class="anchor-link" href="#Protease-dictionary"> </a></h2>
</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>A numba compatible dictionary that stores different regular expressions for proteolytic enzymes. 
The dictionary is identical to the one in <em>AlphaPept</em> and was largely taken from the <em>Pyteomics</em> website which in turn derived the rules are from <em>expasy</em>.</p>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">protease_dict</span><span class="p">[</span><span class="s2">&quot;trypsin&quot;</span><span class="p">]</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_text output_subarea output_execute_result">
<pre>&#39;([KR](?=[^P]))&#39;</pre>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Get-proteolytic-cleavage-sites-for-a-protein-sequence">Get proteolytic cleavage sites for a protein sequence<a class="anchor-link" href="#Get-proteolytic-cleavage-sites-for-a-protein-sequence"> </a></h2>
</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="get_cleavage_sites" class="doc_header"><code>get_cleavage_sites</code><a href="https://github.com/mannlabs/alphamap/tree/master/alphamap/proteolytic_cleavage.py#L49" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>get_cleavage_sites</code>(<strong><code>sequence</code></strong>:<code>str</code>, <strong><code>protease</code></strong>:<code>str</code>)</p>
</blockquote>
<p>Function to get the position of proteolytic cleavage sites in a sequence.</p>
<p>Args:
    sequence (str): Amino acid sequence.
    protease (str): Protease to use for in silico digestion.
Returns:
    list: List of cleavage site indices for the selected protease.</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

</div>
 

