---

title: Sequence visualization


keywords: fastai
sidebar: home_sidebar



nb_path: "nbs/SequencePlot.ipynb"
---
<!--

#################################################
### THIS FILE WAS AUTOGENERATED! DO NOT EDIT! ###
#################################################
# file to edit: nbs/SequencePlot.ipynb
# command to build the docs after a change: nbdev_build_docs

-->

<div class="container" id="notebook-container">
        
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Format-uniprot-annotation">Format uniprot annotation<a class="anchor-link" href="#Format-uniprot-annotation"> </a></h2>
</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="format_uniprot_annotation" class="doc_header"><code>format_uniprot_annotation</code><a href="https://github.com/mannlabs/alphamap/tree/master/alphamap/sequenceplot.py#L15" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>format_uniprot_annotation</code>(<strong><code>uniprot_ann</code></strong>:<code>DataFrame</code>, <strong><code>uniprot_feature_dict</code></strong>:<code>dict</code>)</p>
</blockquote>
<p>Function to format uniprot annotation for plotting.</p>
<p>Args:
    uniprot_ann (pd.DataFrame): Formatted uniprot annotations from alphamap.
    uniprot_feature_dict (dict): Uniprot feature dictionary defined by alphamap.
Returns:
    pd.DataFrame: Uniprot annotation with a combined structure entry for helix, strand and turn.</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Format-data-for-plotting">Format data for plotting<a class="anchor-link" href="#Format-data-for-plotting"> </a></h2>
</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="get_plot_data" class="doc_header"><code>get_plot_data</code><a href="https://github.com/mannlabs/alphamap/tree/master/alphamap/sequenceplot.py#L89" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>get_plot_data</code>(<strong><code>protein</code></strong>, <strong><code>df</code></strong>, <strong><code>fasta</code></strong>)</p>
</blockquote>
<p>Function to format experimental data for plotting.</p>
<p>Args:
    protein (str): Uniprot protein accession.
    df (pd.DataFrame): Experimental data imported and formatted according to alphamap standards.
    fasta (fasta): Fasta file imported by pyteomics 'fasta.IndexedUniProt'.
Returns:
    pd.DataFrame: Formatted dataframe for plotting.</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Function-to-plot-a-single-dataset">Function to plot a single dataset<a class="anchor-link" href="#Function-to-plot-a-single-dataset"> </a></h2>
</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="plot_single_peptide_traces" class="doc_header"><code>plot_single_peptide_traces</code><a href="https://github.com/mannlabs/alphamap/tree/master/alphamap/sequenceplot.py#L181" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>plot_single_peptide_traces</code>(<strong><code>df_plot</code></strong>, <strong><code>protein</code></strong>, <strong><code>fasta</code></strong>)</p>
</blockquote>
<p>Function to plot single peptide trace.</p>
<p>Args:
    df_plot (pd.DataFrame): Formatted dataframe for plotting, generated by get_plot_data.
    protein (str): Uniprot protein accession.
    fasta (fasta): Fasta file imported by pyteomics 'fasta.IndexedUniProt'.
Returns:
    go.Figure: Figure data for a single dataset.</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Plotting-function-for-the-full-sequence-plot">Plotting function for the full sequence plot<a class="anchor-link" href="#Plotting-function-for-the-full-sequence-plot"> </a></h2>
</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="get_quality_category" class="doc_header"><code>get_quality_category</code><a href="https://github.com/mannlabs/alphamap/tree/master/alphamap/sequenceplot.py#L403" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>get_quality_category</code>(<strong><code>s</code></strong>)</p>
</blockquote>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="get_exposure_category" class="doc_header"><code>get_exposure_category</code><a href="https://github.com/mannlabs/alphamap/tree/master/alphamap/sequenceplot.py#L415" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>get_exposure_category</code>(<strong><code>s</code></strong>)</p>
</blockquote>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="get_alphafold_annotation" class="doc_header"><code>get_alphafold_annotation</code><a href="https://github.com/mannlabs/alphamap/tree/master/alphamap/sequenceplot.py#L427" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>get_alphafold_annotation</code>(<strong><code>protein</code></strong>:<code>str</code>, <strong><code>selected_features</code></strong>:<code>list</code>, <strong><code>download_folder</code></strong>:<code>str</code>=<em><code>'/var/folders/zx/n29r0swn0hddt1qgy159sjrw0000gn/T'</code></em>)</p>
</blockquote>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="c1"># af_annotation = get_alphafold_annotation(protein= &quot;Q00266&quot;, </span>
<span class="c1">#                              selected_features= [&quot;AlphaFold confidence&quot;,  </span>
<span class="c1">#                                                    &quot;AlphaFold exposure&quot;, </span>
<span class="c1">#                                                    &quot;AlphaFold IDR&quot;, </span>
<span class="c1">#                                                    &quot;AlphaFold secondary structures&quot;], </span>
<span class="c1">#                              download_folder= &#39;/Users/isabell/Downloads/&#39;) </span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="plot_peptide_traces" class="doc_header"><code>plot_peptide_traces</code><a href="https://github.com/mannlabs/alphamap/tree/master/alphamap/sequenceplot.py#L518" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>plot_peptide_traces</code>(<strong><code>df</code></strong>:<code>DataFrame</code>, <strong><code>name</code></strong>:<code>str</code>, <strong><code>protein</code></strong>:<code>str</code>, <strong><code>fasta</code></strong>:<code>pyteomics.fasta</code>, <strong><code>uniprot</code></strong>:<code>DataFrame</code>, <strong><code>selected_features</code></strong>:<code>list</code>, <strong><code>uniprot_feature_dict</code></strong>:<code>dict</code>, <strong><code>uniprot_color_dict</code></strong>:<code>dict</code>, <strong><code>selected_proteases</code></strong>:<code>list</code>=<em><code>[]</code></em>, <strong><code>selected_alphafold_features</code></strong>:<code>list</code>=<em><code>[]</code></em>, <strong><code>dashboard</code></strong>:<code>bool</code>=<em><code>False</code></em>, <strong><code>trace_colors</code></strong>:<code>list</code>=<em><code>[]</code></em>, <strong><code>download_folder</code></strong>:<code>str</code>=<em><code>'/var/folders/zx/n29r0swn0hddt1qgy159sjrw0000gn/T'</code></em>)</p>
</blockquote>
<p>Function to generate the sequence plot.</p>
<p>Args:
    df (pd.DataFrame/list): Single dataframe or list of dataframes containing the datasets to plot.
    name (str/list): Single string or list of strings containing the names for each dataset in df.
    protein (str): Uniprot protein accession.
    fasta (fasta): Fasta file imported by pyteomics 'fasta.IndexedUniProt'.
    uniprot (pd.DataFrame): Uniprot annotations formatted by alphamap.
    selected_features (list): List of uniprot features to plot.
    uniprot_feature_dict (dict): Uniprot feature dictionary.
    uniprot_color_dict (dict): Uniprot color dictionary.
    selected_proteases (list, optional): List of proteases to plot. Default is an empty list.
    dashboard (bool, optional): Flag if the function is called from the dashboard. Default is 'False'.
    trace_colors (list, optional): List of manualy selected colors for each dataset in df. Default is an empty list.</p>
<p>Returns:
    go.Figure: Sequence plot.</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h3 id="3D-visualization">3D visualization<a class="anchor-link" href="#3D-visualization"> </a></h3>
</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="extract_annotation" class="doc_header"><code>extract_annotation</code><a href="https://github.com/mannlabs/alphamap/tree/master/alphamap/sequenceplot.py#L873" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>extract_annotation</code>(<strong><code>df</code></strong>:<code>DataFrame</code>)</p>
</blockquote>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="manipulate_cif" class="doc_header"><code>manipulate_cif</code><a href="https://github.com/mannlabs/alphamap/tree/master/alphamap/sequenceplot.py#L884" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>manipulate_cif</code>(<strong><code>protein</code></strong>:<code>str</code>, <strong><code>MS_data</code></strong>:<code>DataFrame</code>, <strong><code>download_folder</code></strong>:<code>str</code>=<em><code>'/var/folders/zx/n29r0swn0hddt1qgy159sjrw0000gn/T'</code></em>)</p>
</blockquote>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="adjust_html" class="doc_header"><code>adjust_html</code><a href="https://github.com/mannlabs/alphamap/tree/master/alphamap/sequenceplot.py#L985" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>adjust_html</code>(<strong><code>protein</code></strong>:<code>str</code>, <strong><code>coloring</code></strong>:<code>str</code>, <strong><code>data_name</code></strong>:<code>str</code>)</p>
</blockquote>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="get_ms_concensus" class="doc_header"><code>get_ms_concensus</code><a href="https://github.com/mannlabs/alphamap/tree/master/alphamap/sequenceplot.py#L1102" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>get_ms_concensus</code>(<strong><code>ms_list</code></strong>)</p>
</blockquote>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="plot_3d_structure" class="doc_header"><code>plot_3d_structure</code><a href="https://github.com/mannlabs/alphamap/tree/master/alphamap/sequenceplot.py#L1115" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>plot_3d_structure</code>(<strong><code>df</code></strong>:<code>DataFrame</code>, <strong><code>name</code></strong>:<code>str</code>, <strong><code>protein</code></strong>:<code>str</code>, <strong><code>fasta</code></strong>:<code>pyteomics.fasta</code>, <strong><code>selected_coloring</code></strong>:<code>str</code>, <strong><code>dashboard</code></strong>:<code>bool</code>=<em><code>False</code></em>, <strong><code>download_folder</code></strong>:<code>str</code>=<em><code>'/var/folders/zx/n29r0swn0hddt1qgy159sjrw0000gn/T'</code></em>)</p>
</blockquote>
<p>Function to generate the sequence plot.</p>
<p>Args:
    df (pd.DataFrame/list): Single dataframe or list of dataframes containing the datasets to plot.
    name (str): Single string containing the name of the MS dataset in df.
    protein (str): Uniprot protein accession.
    fasta (fasta): Fasta file imported by pyteomics 'fasta.IndexedUniProt'.
    selected_coloring (str): Coloring to show.
    dashboard (bool, optional): Flag if the function is called from the dashboard. Default is 'False'.</p>
<p>Returns:
    go.Figure: 3D plot.</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="c1"># &#39;alphafold_quality&#39;</span>
<span class="c1"># &#39;alphafold_exposure&#39;</span>
<span class="c1"># &#39;alphafold_IDR&#39;</span>
<span class="c1"># &#39;alphafold_secondary_structures&#39;</span>
<span class="c1"># &#39;MS data&#39;</span>
<span class="c1"># &#39;MS PTMs&#39;</span>
<span class="c1"># &#39;MS amino acids&#39;</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="c1"># plot_3d_structure(df = formatted_diann_data,</span>
<span class="c1">#                     name = &#39;proteome&#39;,</span>
<span class="c1">#                     protein = &quot;P23284&quot;, #Q9H7C9, P23284</span>
<span class="c1">#                     fasta = human_fasta,</span>
<span class="c1">#                     selected_coloring = &#39;AlphaFold quality&#39;, #&#39;AlphaFold exposure&#39;, #&#39;AlphaFold quality&#39;, &#39;MS data&#39;</span>
<span class="c1">#                     dashboard = False)</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Create-a-pdf-report">Create a pdf report<a class="anchor-link" href="#Create-a-pdf-report"> </a></h2>
</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="create_pdf_report" class="doc_header"><code>create_pdf_report</code><a href="https://github.com/mannlabs/alphamap/tree/master/alphamap/sequenceplot.py#L1191" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>create_pdf_report</code>(<strong><code>proteins</code></strong>:<code>list</code>, <strong><code>df</code></strong>:<code>DataFrame</code>, <strong><code>name</code></strong>:<code>str</code>, <strong><code>fasta</code></strong>:<code>pyteomics.fasta</code>, <strong><code>uniprot</code></strong>:<code>DataFrame</code>, <strong><code>selected_features</code></strong>:<code>list</code>, <strong><code>uniprot_feature_dict</code></strong>:<code>dict</code>, <strong><code>uniprot_color_dict</code></strong>:<code>dict</code>, <strong><code>selected_proteases</code></strong>:<code>list</code>=<em><code>[]</code></em>, <strong><code>trace_colors</code></strong>:<code>list</code>=<em><code>[]</code></em>)</p>
</blockquote>
<p>Function to write pdf reports for selected proteins</p>
<p>Args:
    proteins (list): List of uniprot protein accessions.
    df (pd.DataFrame/list): Single dataframe or list of dataframes containing the datasets to plot.
    name (str/list): Single string or list of strings containing the names for each dataset in df.
    fasta (fasta): Fasta file imported by pyteomics 'fasta.IndexedUniProt'.
    uniprot (pd.DataFrame): Uniprot annotations formatted by alphamap.
    selected_features (list): List of uniprot features to plot.
    uniprot_feature_dict (dict): Uniprot feature dictionary.
    uniprot_color_dict (dict): Uniprot color dictionary.
    selected_proteases (list, optional): List of proteases to plot. Default is an empty list.
    trace_colors (list, optional): List of manualy selected colors for each dataset in df. Default is an empty list.</p>
<p>Returns:
    BytesIO: BytesIO object for writing a pdf report.</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

</div>
 

